---
title: "Analysis normalized data"
author: "Ruth Gómez Graciani"
date: "10/5/2020"
output: html_document
editor_options: 
  chunk_output_type: inline
---

```{r, setup, include=FALSE}
# require("knitr")
# require("tidyr")
# require("ggforce")
# require("ggpubr")
# require("reshape2")
library("ggplot2")
# require("stringr")
knitr::opts_knit$set(root.dir = "~/20200401_RealRecombination/")
# outdir<-"report/2020-08-20_defMethods/"


```

# Proof of concept

These files are from analyzing Group 3 inversions in fixed window sizes. If this works, I'll do it with all inversions and different window sizes. 

```{r XrecRate,  echo = FALSE, eval=FALSE}

normData <- read.table("analysis/2020-10-02_09_crossovers_invsG3/crossoverResult_QN.txt", header = TRUE)

```

I also need the genotypes

```{r createGenotypes,  echo = FALSE, eval=FALSE}

 # Individual genotypes
   # Get individuals genotypes
    genotypes<-read.table("analysis/2020-05-23_07_imputationTables/imputationResults.csv", header=TRUE, sep = ",")
  
    # Get individual names
    genotypes$Individual<-sub("^\\d+\\.","",genotypes$Individual)
    genotypes$Individual<-sub("nc", "NC", genotypes$Individual, ignore.case = FALSE)
  
#---------------------GENOTYPE QUALITY CONTROL FUNCTION-----------------------------------
    
makegeno<-function(genotypes){
    genotypes_characterized<-genotypes[genotypes$Population %in% c("EUR", "AFR"),]
    
    eur.quality<-genotypes_characterized[genotypes_characterized$Population == "EUR" & genotypes_characterized$EUR %in% c("Tagged", "Imputable") ,]
    afr.quality<-genotypes_characterized[genotypes_characterized$Population == "AFR" & genotypes_characterized$AFR %in% c("Tagged", "Imputable") ,]
    
    perc.good.uni.eur<-table(eur.quality$Uniformity)/sum(table(eur.quality$Uniformity))*100
    perc.good.uni.afr<-table(afr.quality$Uniformity)/sum(table(afr.quality$Uniformity))*100
    
    
    testvec.eur<-eur.quality[ eur.quality$Quality %in% "good",] 
    testvec.afr<-afr.quality[ afr.quality$Quality %in% "good", ] 
    
    perc.good.uni.eur.c<-table(  testvec.eur$Uniformity)
    perc.good.uni.afr.c<-table( testvec.afr$Uniformity)
    
    aresame.eur<-eur.quality[ eur.quality$Quality %in% "good" & eur.quality$Uniformity %in% c("alert", "bad"), ] 
    aresame.afr<-afr.quality[ afr.quality$Quality %in% "good" & afr.quality$Uniformity %in% c("alert", "bad"), ] 

    # En europeus la inversio 191 dona problemes en 5/12(Es la unica) i en africans la 1075(1/3). Aquestes dues es prendran amb pinces. La resta, si la qualitat es bona en general el resultat tambe. 
    
    accepted.eur<-genotypes[genotypes$Population == "EUR" & genotypes$EUR %in% c("Tagged", "Imputable") & genotypes$Probability_con >0.8,]
    accepted.afr<-genotypes[genotypes$Population == "AFR" & genotypes$AFR %in% c("Tagged", "Imputable") & genotypes$Probability_con >0.8,]

  # SAS i ALL no hi ha moltes referencies
    test.sas<-genotypes[ (genotypes$GLB %in% c("Tagged", "Imputable") 
                          # | genotypes$AFR %in% c("Tagged", "Imputable") 
                          # | genotypes$EUR %in% c("Tagged", "Imputable") 
                          ) 
                         & genotypes$Population %in% c("ALL", "SAS") , ]
    test.sas[test.sas$Genotype_500 == "?", "Genotype_500"]<-NA
    test.sas$Genotype_500<-as.factor(as.character(test.sas$Genotype_500))
    test.sas$accepted<-"no"
    test.sas[test.sas$Genotype_100 == test.sas$Genotype_250 & test.sas$Genotype_250 == test.sas$Genotype_500 & 
             test.sas$Probability_100 > 0.8 & test.sas$Probability_250 > 0.8 & test.sas$Probability_500 > 0.8, "accepted" ]<-"yes"
    test.sas[test.sas$Inversion %in% c("HsInv0191", "HsInv1075"),"accepted"]<-"no"
    test.sas[test.sas$Uniformity == "good" & test.sas$Quality == "good","accepted"]<-"yes"
    
    accepted.sasall<-test.sas[test.sas$accepted == "yes",]
    accepted.sasall$accepted <- NULL
    
  # Asiatics 
     accepted.eas<-genotypes[genotypes$Population == "EAS" & 
                               (genotypes$GLB %in% c("Tagged", "Imputable") 
                                # | genotypes$AFR %in% c("Tagged", "Imputable")
                                # | genotypes$EUR %in% c("Tagged", "Imputable") 
                                )
                             & genotypes$Uniformity == "good" & genotypes$Quality %in% c("good", "alert") & genotypes$Probability_con > 0.8,]
   

     accepted.genotypes<-rbind(accepted.afr, accepted.eas, accepted.eur, accepted.sasall)
 return(accepted.genotypes)
    }
    
 # ------------------------------------------------------------
    
    genotypes<-makegeno(genotypes)
    
    # length(unique(genotypes$Inversion))
```

```{r addGenotypes,  echo = FALSE, eval=FALSE}    
    # Merge with info
    normData<-  merge(normData, genotypes[,c("Individual", "Inversion", "Genotype_500")], by.x = c("ind", "inv") , by.y = c("Individual", "Inversion" ), all.x = TRUE)
    normData$Haplotype<-NA
    normData[normData$Genotype_500 %in% c("INV", "STD"), "Haplotype"]<-"Homozygous"
    normData[normData$Genotype_500 %in% c("HET"), "Haplotype"]<-"Heterozygous"
    
    # Clean up
    # rm(genotypes)

```

Now I want to make the summary table that aggregates by genotpe

```{r aggregateTable ,  echo = FALSE, eval=FALSE}

# Parse position markers
  normData$position <- factor(normData$position , levels = c("left", "inv", "right"))
  normData$posNumber<- as.numeric(str_split_fixed(normData$idWin, "_", 3)[,3])

# Order table and factors
  normData<-normData[order( normData$inv, normData$position, normData$posNumber),]
  normData$idWin<- factor(normData$idWin, levels = unique(normData$idWin))
```

```{r PlotFunction ,  echo = FALSE, eval=FALSE}
# Function to aggregate values by inversion
    function.InvPlot.means<-function(inversion, toPlot){
 
    test.subset<-aggregate(   cM.Mb.QN  ~  inv + startWin+endWin + Haplotype +idWin+position,toPlot[toPlot$inv == inversion,], mean)
    
toPlot$counts<-1
test.counts<-aggregate(   counts  ~  inv + startWin+endWin + Haplotype +idWin+position,toPlot[toPlot$inv == inversion,], sum)
test.subset<-merge(test.subset, test.counts)


     a<-      ggplot(test.subset)+geom_rect(aes(xmin=startWin, xmax=endWin, ymin = 0, ymax = cM.Mb.QN  , fill = counts)) +
      facet_grid(. ~ Haplotype, scales = "free") +
      geom_rect(data = test.subset[test.subset$position=="inv",], aes(xmin=startWin, xmax=endWin, ymin = 0, ymax = cM.Mb.QN), fill="goldenrod")+
       ggtitle(inversion) #+
       # theme(strip.text.x = element_text(size = 20), axis.text.y = element_text(size= 20),axis.text.x = element_text(size= 15),title = element_text(size=20))
    return(a)


    }
  
```

```{r makePlotlist ,  echo = FALSE, eval=FALSE}
    toPlot<-normData[!is.na(normData$Haplotype),]
    plotlist<-list()
 
  for(i in unique(toPlot$inv)){
    
   if (length( table(toPlot[toPlot$inv == i,"Haplotype"])) > 1){
    plotlist[[i]]<-function.InvPlot.means(i, toPlot )
   }
  }

  # plotname<-paste0(outdir, paste0("images/invDescription_all",".png"))
  # png(filename=plotname, width = 3508/2, height = 2480/2, units = "px" )
  plot<- annotate_figure( ggarrange(plotlist=plotlist, common.legend = TRUE), top = text_grob(paste0("Recombination rate percentiles aggregated by zygosity" ), size = 50) )
  # print(plot)
  # dev.off()
  
  table(toPlot$inv, toPlot$Haplotype, toPlot$ind)
  
inversion <- "HsInv1110"
plotlist$HsInv0286
plotlist$HsInv0290
plotlist$HsInv0786
```


The result seems promising!!

# All inversions analysis

Now that I know the kind of plot that we need, I'll repeat et for all inversions and 3 window sizes: 10kb, 50kb, 200kb. I want to make a function or group of functions to streamline the process. 


```{r allProcessFun,  echo = FALSE, eval=FALSE}

plotInvsQN<-function(normData, genotypes=genotypes){

    # Add genotypes to table
    normData<-  merge(normData, genotypes[,c("Individual", "Inversion", "Genotype_500")], by.x = c("ind", "inv") , by.y = c("Individual", "Inversion" ), all.x = TRUE)
    normData$Haplotype<-NA
    normData[normData$Genotype_500 %in% c("INV", "STD"), "Haplotype"]<-"Homozygous"
    normData[normData$Genotype_500 %in% c("HET"), "Haplotype"]<-"Heterozygous"
    
    # Parse position markers
    normData$position <- factor(normData$position , levels = c("left", "inv", "right"))
    normData$posNumber<- as.numeric(str_split_fixed(normData$idWin, "_", 3)[,3])

  # Order table and factors
    normData<-normData[order( normData$inv, normData$position, normData$posNumber),]
    normData$idWin<- factor(normData$idWin, levels = unique(normData$idWin))
    
  # Make plottable inversions list
    toPlot<-normData[!is.na(normData$Haplotype),]
    
  # Make empty plot list
    plotlist<-list()
 
  for(i in unique(toPlot$inv)){
    
   if (length( table(toPlot[toPlot$inv == i,"Haplotype"])) > 1){
    plotlist[[i]]<-function.InvPlot.means(i, toPlot )
   }
  }
  
  return(plotlist)  
  
}

```

```{r analyze200,  echo = FALSE, eval=FALSE}

normData200 <- read.table("analysis/2020-10-14_09_crossovers_200k/crossoverResult_QN.txt", header = TRUE)

# Make sure you have the 'genotypes' table!! It is created in the Create Genotypes chunk!
# In addition, this function uses function.InvPlot.means from PlotFunction chunk

plotlist.200<- plotInvsQN(normData200, genotypes)

# Discard inversions that we know that won't work!
plotlist.200$HsInv0124<-NULL

# Now different plots can be called
  # dir.create("report/2020-10-05_analysisNormalizedData/images")
  
  # printplot
  plotname<- "report/2020-10-05_analysisNormalizedData/images/allinvs.b.200k.png"
  png(filename=plotname, width = 3508, height = 2480, units = "px" )
  plot<- annotate_figure( ggarrange(plotlist=plotlist.200, common.legend = TRUE), top = text_grob(paste0("Recombination rate percentiles aggregated by zygosity" ), size = 50) )
  print(plot)
  dev.off()

```

```{r analyze50,  echo = FALSE, eval=FALSE}

normData50 <- read.table("analysis/2020-10-14_09_crossovers_50k/crossoverResult_QN.txt", header = TRUE)

# Make sure you have the 'genotypes' table!! It is created in the Create Genotypes chunk!
# In addition, this function uses function.InvPlot.means from PlotFunction chunk

plotlist.50<- plotInvsQN(normData50, genotypes)

# Discard inversions that we know that won't work!
plotlist.50$HsInv0124<-NULL
plotlist.50$HsInv1110<-NULL

# Now different plots can be called
  # dir.create("report/2020-10-05_analysisNormalizedData/images")
  
  # printplot
  plotname<- "report/2020-10-05_analysisNormalizedData/images/allinvs.50k.png"
  png(filename=plotname, width = 3508, height = 2480, units = "px" )
  plot<- annotate_figure( ggarrange(plotlist=plotlist.50, common.legend = TRUE), top = text_grob(paste0("Recombination rate percentiles aggregated by zygosity" ), size = 50) )
  print(plot)
  dev.off()


```

```{r analyze10,  echo = FALSE, eval=FALSE}

normData10 <- read.table("analysis/2020-10-15_09_crossovers_10k/crossoverResult_QN.txt", header = TRUE)

# Make sure you have the 'genotypes' table!! It is created in the Create Genotypes chunk!
# In addition, this function uses function.InvPlot.means from PlotFunction chunk

plotlist.10<- plotInvsQN(normData10, genotypes)

# Discard inversions that we know that won't work!
plotlist.10$HsInv0124<-NULL
plotlist.10$HsInv1110<-NULL

# Now different plots can be called
  # dir.create("report/2020-10-05_analysisNormalizedData/images")
  
  # printplot
  plotname<- "report/2020-10-05_analysisNormalizedData/images/allinvs.10k.png"
  png(filename=plotname, width = 3508, height = 2480, units = "px" )
  plot<- annotate_figure( ggarrange(plotlist=plotlist.10, common.legend = TRUE), top = text_grob(paste0("Recombination rate percentiles aggregated by zygosity" ), size = 50) )
  print(plot)
  dev.off()


```


Well that works fine!! But we need a way to see general results, e.g. the difference between left and right. Since all regions are in the same scale, absolute difference should work fine. In addition, I want to know the power (how many het vs homo individuals I had) and the size of the inversion. 

```{r allProcessDifFun }

plotInvsDifQN<-function(normData, genotypes=genotypes){

   # Add genotypes to table
    normData<-  merge(normData, genotypes[,c("Individual", "Inversion", "Genotype_500")], by.x = c("ind", "inv") , by.y = c("Individual", "Inversion" ), all.x = TRUE)
    normData$Haplotype<-NA
    normData[normData$Genotype_500 %in% c("INV", "STD"), "Haplotype"]<-"Homozygous"
    normData[normData$Genotype_500 %in% c("HET"), "Haplotype"]<-"Heterozygous"
    
  # Make plottable inversions list
    toPlot<-normData[!is.na(normData$Haplotype),]
    
      test.subset<-aggregate(   cM.Mb.QN  ~  inv + startWin+endWin + Haplotype +idWin,toPlot, mean)
    
toPlot$counts<-1
test.counts<-aggregate(   counts  ~  inv + startWin+endWin + Haplotype +idWin,toPlot, sum)
test.subset<-merge(test.subset, test.counts)


      test.subset.counts<-dcast(test.subset, inv + startWin + endWin + idWin  ~ Haplotype, value.var="counts")
      colnames(test.subset.counts)[colnames(test.subset.counts) %in% c( "Heterozygous", "Homozygous"  )]<-c("Heterozygous.counts", "Homozygous.counts"  )
      test.subset.values<-dcast(test.subset, inv + startWin + endWin + idWin  ~ Haplotype, value.var="cM.Mb.QN")   
      colnames(test.subset.values)[colnames(test.subset.values) %in% c( "Heterozygous", "Homozygous"  )]<-c("Heterozygous.cM.Mb.QN", "Homozygous.counts.cM.Mb.QN" )
      
      test.subset.wide<-merge(test.subset.counts, test.subset.values)
      
      test.subset.wide$Means.Difference<-test.subset.wide$Homozygous.counts.cM.Mb.QN - test.subset.wide$Heterozygous.cM.Mb.QN
      test.subset.wide$Power.score<-log(test.subset.wide$Homozygous.counts / test.subset.wide$Heterozygous.counts)
      
      # Discard those without data
      test.subset.wide<-  test.subset.wide[!is.na(test.subset.wide$Means.Difference),]
       # test.subset.wide<- test.subset.wide[!(test.subset.wide$inv %in% c("HsInv0290", "HsInv0786")),]
      # Order table and factors
      
      test.subset.wide[c("side", "relpos")]<-str_split_fixed(test.subset.wide$idWin, "_", 3)[,c(2,3)]
      test.subset.wide[c("winPos")]<-str_split_fixed(test.subset.wide$idWin, "_", 2)[,2]
      

      test.subset.wide$side<-factor(test.subset.wide$side, levels = c("left", "inv", "right"))
      test.subset.wide$relpos<-as.numeric(test.subset.wide$relpos)
      
      test.subset.wide<-test.subset.wide[order( test.subset.wide$side, test.subset.wide$relpos),]
       test.subset.wide$winPos<- factor(  test.subset.wide$winPos, levels = unique(  test.subset.wide$winPos))
          
   
      invinfo<-read.table("data/use/inversions_info/2020.07.inversions_hg38.gff")
      invinfo$size<-invinfo$V5-invinfo$V4
      invinfo$Group<-"Small"
      invinfo[invinfo$size > 5000,"Group"]<-"Medium"
      invinfo[invinfo$size > 50000,"Group"]<-"Big"
      
      invinfo$inv<-gsub("ID=1;Name=", "", invinfo$V9)
      invinfo$inv<-gsub(";.*$", "", invinfo$inv)
      
      test.subset.wide<-merge(test.subset.wide, invinfo[c("inv", "Group")], all.x = TRUE)
      
       plot<-ggplot(test.subset.wide)+geom_line(aes(x = winPos, y = Means.Difference, group = inv, color = Power.score))+
         facet_grid(Group ~ . )+
         geom_text(data = test.subset.wide[test.subset.wide$winPos == "inv_1", ] , aes(x = winPos, y = Means.Difference, label = inv)  )
 
       
      # return(plot)
       return(test.subset.wide)
}
      
```


```{r analyzed200,  echo = FALSE, eval=FALSE}

normData200 <- read.table("analysis/2020-10-14_09_crossovers_200k/crossoverResult_QN.txt", header = TRUE)

# Make sure you have the 'genotypes' table!! It is created in the Create Genotypes chunk!
# In addition, this function uses function.InvPlot.means from PlotFunction chunk

plotdif.200<- plotInvsDifQN(normData200, genotypes)

# Now different plots can be called

  # printplot
  plotname<- "report/2020-10-05_analysisNormalizedData/images/allinvs.diffs.200k.nowin.png"
  png(filename=plotname, width = 3508, height = 2480, units = "px" )
  print( plotdif.200)
  dev.off()

 plotdif.200
```



```{r analyzed50,  echo = FALSE, eval=FALSE}

normData50 <- read.table("analysis/2020-10-14_09_crossovers_50k/crossoverResult_QN.txt", header = TRUE)

# Make sure you have the 'genotypes' table!! It is created in the Create Genotypes chunk!
# In addition, this function uses function.InvPlot.means from PlotFunction chunk

plotdif.50<- plotInvsDifQN(normData50, genotypes)

# Now different plots can be called

  # printplot
  plotname<- "report/2020-10-05_analysisNormalizedData/images/allinvs.diffs.50k.png"
  png(filename=plotname, width = 3508, height = 2480, units = "px" )
  print( plotdif.50)
  dev.off()

 plotdif.50
```



```{r analyzed50,  echo = FALSE, eval=FALSE}

normData10 <- read.table("analysis/2020-10-15_09_crossovers_10k/crossoverResult_QN.txt", header = TRUE)

# Make sure you have the 'genotypes' table!! It is created in the Create Genotypes chunk!
# In addition, this function uses function.InvPlot.means from PlotFunction chunk

plotdif.10<- plotInvsDifQN(normData10, genotypes)

# Now different plots can be called

  # printplot
  plotname<- "report/2020-10-05_analysisNormalizedData/images/allinvs.diffs.10k.png"
  png(filename=plotname, width = 3508, height = 2480, units = "px" )
  print( plotdif.10)
  dev.off()

 # plotdif.50
```


Now, these plots have a lot of noise. When we see the signal, what we see is "most of big inversions have less rec.rate inside". So, I'll count these situations and make a summary plot


```{r summaryplot,  echo = FALSE, eval=FALSE}

mydata<-plotInvsDifQN(normData10, genotypes)


# I want the mean difference across windows in a zone

plot.table<-aggregate(   Means.Difference  ~  inv + side+Group, mydata,mean)
plot.table<-plot.table[!(plot.table$inv %in% c("HsInv1110", "HsInv0124")),]

ggplot(plot.table)+geom_histogram(aes( x = Means.Difference))+facet_grid(side  ~. )

plot.table$difpos <-"0"
plot.table[plot.table$Means.Difference < 0,"difpos"]<-"-"
plot.table[plot.table$Means.Difference > 0,"difpos"]<-"+"

ggplot(plot.table)+geom_histogram(aes( x = difpos), stat="count")+facet_grid(side  ~Group )


#  Somep lots for the presentation

mydata<-plotInvsDifQN(normData200, genotypes)

mydata<-mydata[mydata$inv == "HsInv0286",]

ggplot(mydata)+geom_bar(aes(x = winPos, y = Means.Difference), stat="identity")

pt<-plot.table[!duplicated(plot.table$inv),]
table(pt$Group)
```


# Normalization confirmation




```{r summaryplot,  echo = FALSE, eval=FALSE}
 
map.200.nonorm<-read.table("data/use/recMaps/recMap_200000.txt", header = TRUE)

newstats.nonorm <- aggregate(cM.Mb ~  ind ,map.200.nonorm, mean)

ggplot(map.200.nonorm[map.200.nonorm$ind %in% c("NC1", "NC2", "NC3", "NC4"),], aes(x = cM.Mb ,  fill = ind))+geom_histogram(bins = 100)+facet_grid(ind ~ .)+scale_x_log10()



map.200<-read.table("analysis/2020-10-15_09_crossovers_allMap200/crossoverResult_QN.txt", header = TRUE) 

head(map.200)


ggplot(map.200)+geom_boxplot(aes(x = ind, y = cM.Mb.QN))

newstats <- aggregate(cM.Mb.QN ~  ind ,map.200, mean)

ggplot(map.200, aes(x = ind, y = cM.Mb.QN, fill = ind) )+geom_boxplot()+
  # stat_summary(fun.y=mean, geom="point", color = "red", shape = 3, size = 10) + theme(legend.position = "none")+
  # geom_text(data = newstats, aes(x = variable, y = value.scaled, label = value.scaled), nudge_y = 0.15)+
    geom_text(data = newstats, aes(x = ind, y = cM.Mb.QN, label = round(cM.Mb.QN, 10)), nudge_y =0.1)


ggplot(map.200[map.200$ind %in% c("NC1", "NC2", "NC3", "NC4"),], aes(x = cM.Mb.QN ,  fill = ind))+geom_histogram(bins = 100)+facet_grid(ind ~ .)
 
# ----------------

sample<-map.200.nonorm[, c("cM.Mb", "ind", "idWin")]
sample_wide<-dcast(sample,  idWin ~ ind , value.var = "cM.Mb")

quantile_normalisation <- function(df){
  df_rank <- apply(df,2,rank,ties.method="min")
  df_sorted <- data.frame(apply(df, 2, sort))
  df_mean <- apply(df_sorted, 1, mean)
   
  index_to_mean <- function(my_index, my_mean){
    return(my_mean[my_index])
  }
   
  df_final <- apply(df_rank, 2, index_to_mean, my_mean=df_mean)
  rownames(df_final) <- rownames(df)
  return(df_final)
}

norm<-quantile_normalisation(sample_wide[,c(2:ncol(sample_wide))])

sample<-melt(data.frame(norm))
ggplot(sample, aes(x = value ,  fill = variable))+geom_histogram(bins = 100)+facet_grid(variable ~ .)#+scale_x_log10()

newstats <- aggregate(value ~  variable ,sample, mean)


# ---- made with python
newstats <- aggregate(cM.Mb.QN ~  ind ,pynorm, mean)
pynorm<-read.table("tmp/crossoverResult.new_QN.txt", header = TRUE)

ggplot(pynorm, aes(x = cM.Mb.QN ,  fill = ind))+geom_histogram(bins = 100)+facet_grid(ind ~ .)#+scale_x_log10()

```





# Some adjustements

## 1: Genotypes

What happened with invs 1256 abd 1590? 

Original inv 1256 has 3het + 17std; original inv 1590 has 5het + 15std. 

```{r   echo = FALSE, eval=FALSE}

 # Individual genotypes
   # Get individuals genotypes
    genotypes<-read.table("analysis/2020-05-23_07_imputationTables/imputationResults.csv", header=TRUE, sep = ",")
  
    # Get individual names
    genotypes$Individual<-sub("^\\d+\\.","",genotypes$Individual)
    genotypes$Individual<-sub("nc", "NC", genotypes$Individual, ignore.case = FALSE)
  
    genotypes<-makegeno(genotypes)
    
    length(unique(genotypes$Inversion))
    nrow(counts)
    counts<-data.frame(unclass(table(as.character(genotypes$Inversion), genotypes$Genotype_500)))

    counts$X. <- NULL
    
    counts$HOMO <- counts$INV + counts$STD
    
    counts$all<-counts$HET + counts$HOMO
    
   nrow(counts[counts$all > 3 & (counts$HET > 0 & counts$HOMO > 0),])
    # length(unique(genotypes$Inversion))
    
    
nDc<- nrow(normData.nodup[normData.nodup$Heterozygous.counts+normData.nodup$Homozygous.counts > 3,]
```



## 2: Preparar dades

```{r createGenotypesGOOD,  echo = FALSE, eval=FALSE}

 # Individual genotypes
   # Get individuals genotypes
    genotypes<-read.table("analysis/2020-05-23_07_imputationTables/imputationResults.csv", header=TRUE, sep = ",")
  
    # Get individual names
    genotypes$Individual<-sub("^\\d+\\.","",genotypes$Individual)
    genotypes$Individual<-sub("nc", "NC", genotypes$Individual, ignore.case = FALSE)
    genotypes$Individual<-sub("ab", "", genotypes$Individual, ignore.case = FALSE)
#---------------------GENOTYPE QUALITY CONTROL FUNCTION-----------------------------------
    
makegeno<-function(genotypes){
    genotypes_characterized<-genotypes[genotypes$Population %in% c("EUR", "AFR"),]
    
    eur.quality<-genotypes_characterized[genotypes_characterized$Population == "EUR" & genotypes_characterized$EUR %in% c("Tagged", "Imputable") ,]
    afr.quality<-genotypes_characterized[genotypes_characterized$Population == "AFR" & genotypes_characterized$AFR %in% c("Tagged", "Imputable") ,]
    
    perc.good.uni.eur<-table(eur.quality$Uniformity)/sum(table(eur.quality$Uniformity))*100
    perc.good.uni.afr<-table(afr.quality$Uniformity)/sum(table(afr.quality$Uniformity))*100
    
    
    testvec.eur<-eur.quality[ eur.quality$Quality %in% "good",] 
    testvec.afr<-afr.quality[ afr.quality$Quality %in% "good", ] 
    
    perc.good.uni.eur.c<-table(  testvec.eur$Uniformity)
    perc.good.uni.afr.c<-table( testvec.afr$Uniformity)
    
    aresame.eur<-eur.quality[ eur.quality$Quality %in% "good" & eur.quality$Uniformity %in% c("alert", "bad"), ] 
    aresame.afr<-afr.quality[ afr.quality$Quality %in% "good" & afr.quality$Uniformity %in% c("alert", "bad"), ] 

    # En europeus la inversio 191 dona problemes en 5/12(Es la unica) i en africans la 1075(1/3). Aquestes dues es prendran amb pinces. La resta, si la qualitat es bona en general el resultat tambe. 
    
    accepted.eur<-genotypes[genotypes$Population == "EUR" & genotypes$EUR %in% c("Tagged", "Imputable") & genotypes$Probability_con >0.8,]
    accepted.afr<-genotypes[genotypes$Population == "AFR" & genotypes$AFR %in% c("Tagged", "Imputable") & genotypes$Probability_con >0.8,]

  # SAS i ALL no hi ha moltes referencies
    test.sas<-genotypes[ (genotypes$GLB %in% c("Tagged", "Imputable") 
                          # | genotypes$AFR %in% c("Tagged", "Imputable") 
                          # | genotypes$EUR %in% c("Tagged", "Imputable") 
                          ) 
                         & genotypes$Population %in% c("ALL", "SAS") , ]
    test.sas[test.sas$Genotype_500 == "?", "Genotype_500"]<-NA
    test.sas$Genotype_500<-as.factor(as.character(test.sas$Genotype_500))
    test.sas$accepted<-"no"
    test.sas[test.sas$Genotype_100 == test.sas$Genotype_250 & test.sas$Genotype_250 == test.sas$Genotype_500 & 
             test.sas$Probability_100 > 0.8 & test.sas$Probability_250 > 0.8 & test.sas$Probability_500 > 0.8, "accepted" ]<-"yes"
    test.sas[test.sas$Inversion %in% c("HsInv0191", "HsInv1075"),"accepted"]<-"no"
    test.sas[test.sas$Uniformity == "good" & test.sas$Quality == "good","accepted"]<-"yes"
    
    accepted.sasall<-test.sas[test.sas$accepted == "yes",]
    accepted.sasall$accepted <- NULL
    
  # Asiatics 
     accepted.eas<-genotypes[genotypes$Population == "EAS" & 
                               (genotypes$GLB %in% c("Tagged", "Imputable") 
                                # | genotypes$AFR %in% c("Tagged", "Imputable")
                                # | genotypes$EUR %in% c("Tagged", "Imputable") 
                                )
                             & genotypes$Uniformity == "good" & genotypes$Quality %in% c("good", "alert") & genotypes$Probability_con > 0.8,]
   

     genotypes<-rbind(accepted.afr, accepted.eas, accepted.eur, accepted.sasall)
     
    # FILTERING GENOTYPES
    # length(unique(genotypes$Inversion))  # 79
    counts<-data.frame(unclass(table(as.character(genotypes$Inversion), genotypes$Genotype_500)))
    counts$X. <- NULL
    counts$HOMO <- counts$INV + counts$STD
    counts$all<-counts$HET + counts$HOMO
    
    counts<-counts[counts$all > 3 & (counts$HET > 0 & counts$HOMO > 0),]
    nrow(counts)
    
    genotypes<-genotypes[genotypes$Inversion %in% rownames(counts),]
    
    return(genotypes)
 }
    
 # ------------------------------------------------------------
    
    genotypes<-makegeno(genotypes)
    


```

```{r PlotFunction2 ,  echo = FALSE, eval=FALSE}
# # Function to aggregate values by inversion
#     function.InvPlot.means<-function(inversion, toPlot){
#      test.subset<-toPlot[toPlot$inv == inversion,]
#     
#      test.melted<-melt(test.subset, id.vars =c("inv" , "startWin","endWin" , "winPos"), measure.vars = c("Homozygous.counts.cM.Mb.QN"  ,"Heterozygous.cM.Mb.QN"))
#      
#      test.agg<-aggregate(   value  ~  inv + startWin+endWin + variable +winPos, test.melted, mean)
#     
#      test.melted<-melt(test.subset, id.vars =c("inv" , "startWin","endWin" , "winPos"), measure.vars = c("Homozygous.counts"  ,"Heterozygous.counts"))
#      
#     colnames(test.agg)[c(4,6)]<-c("Haplotype", "Mean.Fold.Change")
#     levels(test.agg$Haplotype)<-c("Homozygous", "Heterozygous")
#     colnames(test.melted)[c(5,6)]<-c("Haplotype", "Sample.size")
#     levels(test.melted$Haplotype)<-c("Homozygous", "Heterozygous")
# 
#     test.subset<-merge(test.agg, test.melted)
#     test.subset$winPos <- gsub("_\\d", "", test.subset$winPos)
#     
#     test.subset$Mean.Fold.Change.log2<- log2(test.subset$Mean.Fold.Change) 
# 
#      a<-      ggplot(test.subset)+geom_rect(aes(xmin=startWin, xmax=endWin, ymin = 0, ymax = Mean.Fold.Change  , fill = Sample.size)) +
#       facet_grid(. ~ Haplotype, scales = "free") +
#       geom_rect(data = test.subset[test.subset$winPos == "inv",], aes(xmin=startWin, xmax=endWin, ymin = 0, ymax = Mean.Fold.Change), fill="goldenrod")+
#        ggtitle(inversion) #+
#        # theme(strip.text.x = element_text(size = 20), axis.text.y = element_text(size= 20),axis.text.x = element_text(size= 15),title = element_text(size=20))
#     return(a)
# 
# 
#     }
  
```

Ara modificaré la funció que em dona la taula per ajustar varies coses. 

```{r allProcessDifFun }
plotInvsDifQN<-function(normData, genotypes=genotypes, aggregateWins =FALSE){
    # unique(genotypes$Inversion)  # 63
   # Add genotypes to table
    normData<-  merge(normData, genotypes[,c("Individual", "Inversion", "Genotype_500")], by.x = c("ind", "inv") , by.y = c("Individual", "Inversion" ), all.x = TRUE)
    
    normData$Haplotype<-NA
    normData[normData$Genotype_500 %in% c("INV", "STD"), "Haplotype"]<-"Homozygous"
    normData[normData$Genotype_500 %in% c("HET"), "Haplotype"]<-"Heterozygous"
    
    # Make plottable inversions list
    toPlot<-normData[!is.na(normData$Haplotype),]
   
    # Make new summary inside inversion line
    if (aggregateWins == TRUE){
      partn<-aggregate(cM.Mb.QN  ~  inv + Haplotype +position+ind ,  toPlot[toPlot$position == "inv",], mean)
      parts<-aggregate(startWin  ~  inv + Haplotype +position+ind ,  toPlot[toPlot$position == "inv",], min)
      parte<-aggregate(endWin  ~  inv + Haplotype +position+ind ,  toPlot[toPlot$position == "inv",], max)
      part<-merge(partn, merge(parts, parte))
      part$idWin<-paste(part$inv, "inv_1", sep="_" )
      
      toPlot<-rbind(toPlot[toPlot$position != "inv", colnames(part)], part)
    }
    
    # Aggregate by zygosity to make the two sides of the comparison
      # Mean value
      test.subset<-aggregate(   cM.Mb.QN  ~  inv + startWin+endWin + Haplotype +idWin,toPlot, mean)
      # Count of individuals
      toPlot$counts<-1
      test.counts<-aggregate(   counts  ~  inv + startWin+endWin + Haplotype +idWin,toPlot, sum)
      test.subset<-merge(test.subset, test.counts)
    
    # Make both tables wide
      # Counts table wide
      test.subset.counts<-dcast(test.subset, inv + startWin + endWin + idWin  ~ Haplotype, value.var="counts")
      colnames(test.subset.counts)[colnames(test.subset.counts) %in% c( "Heterozygous", "Homozygous"  )]<-c("Heterozygous.counts", "Homozygous.counts"  )
      # CM.Mb table wide
      test.subset.values<-dcast(test.subset, inv + startWin + endWin + idWin  ~ Haplotype, value.var="cM.Mb.QN")   
      colnames(test.subset.values)[colnames(test.subset.values) %in% c( "Heterozygous", "Homozygous"  )]<-c("Heterozygous.cM.Mb.QN", "Homozygous.cM.Mb.QN" )
      
    # Merge wide table 
      test.subset.wide<-merge(test.subset.counts, test.subset.values)
      
      # Filter by zygosity minimum threshold
      
      test.subset.wide<-test.subset.wide [test.subset.wide$Heterozygous.counts > 0 & test.subset.wide$Homozygous.counts > 0,]
      test.subset.wide<-test.subset.wide[(test.subset.wide$Heterozygous.counts + test.subset.wide$Homozygous.counts ) > 3,]

      # Make comparisons
        # Difference
        test.subset.wide$Means.Difference<- test.subset.wide$Heterozygous.cM.Mb.QN - test.subset.wide$Homozygous.cM.Mb.QN 
        # Power ratio
        test.subset.wide$Power.score<-log(test.subset.wide$Homozygous.counts / test.subset.wide$Heterozygous.counts)
        # Fold changes
        test.subset.wide$Means.Fold.Change<- test.subset.wide$Heterozygous.cM.Mb.QN / test.subset.wide$Homozygous.cM.Mb.QN 
        # Fold changes log
        test.subset.wide$Means.Fold.Changelog<- log2(test.subset.wide$Means.Fold.Change)
      
      
      # Discard those without data
      test.subset.wide<-  test.subset.wide[!is.na(test.subset.wide$Means.Difference),]
       # test.subset.wide<- test.subset.wide[!(test.subset.wide$inv %in% c("HsInv0290", "HsInv0786")),]
     
       # Order table and factors
      
      test.subset.wide[c("side", "relpos")]<-str_split_fixed(test.subset.wide$idWin, "_", 3)[,c(2,3)]
      test.subset.wide[c("winPos")]<-str_split_fixed(test.subset.wide$idWin, "_", 2)[,2]
      

      test.subset.wide$side<-factor(test.subset.wide$side, levels = c("left", "inv", "right"))
      test.subset.wide$relpos<-as.numeric(test.subset.wide$relpos)
      
      test.subset.wide<-test.subset.wide[order( test.subset.wide$side, test.subset.wide$relpos),]
       test.subset.wide$winPos<- factor(  test.subset.wide$winPos, levels = unique(  test.subset.wide$winPos))
          
   
      invinfo<-read.table("data/use/inversions_info/2020.07.inversions_hg38.gff")
      invinfo$size<-invinfo$V5-invinfo$V4
      invinfo$Group<-"Small"
      invinfo[invinfo$size > 5000,"Group"]<-"Medium"
      invinfo[invinfo$size > 50000,"Group"]<-"Big"
      
      invinfo$inv<-gsub("ID=1;Name=", "", invinfo$V9)
      invinfo$inv<-gsub(";.*$", "", invinfo$inv)
      
      test.subset.wide<-merge(test.subset.wide, invinfo[c("inv", "Group", "size")], all.x = TRUE)
      
      return(test.subset.wide)
}


# 
#        plot<-ggplot(test.subset.wide)+geom_line(aes(x = winPos, y = Means.Difference, group = inv, color = Power.score))+
#          facet_grid(Group ~ . )+
#          geom_text(data = test.subset.wide[test.subset.wide$winPos == "inv_1", ] , aes(x = winPos, y = Means.Difference, label = inv)  )
#  
#        
      # return(plot)
       
# }
      
```

## 3: Repetir plots que ja existeixen
I repeteixo el plot de les línies

```{r}

# 200 KB

normData200 <- read.table("analysis/2020-10-14_09_crossovers_200k/crossoverResult_QN.txt", header = TRUE)

# Segona versió dels plots de línies, per grups, fent servir mitjanes de les finestres interiors de les invs i fold changes
normData200.lineplot <- plotInvsDifQN(normData200, genotypes=genotypes, aggregateWins = TRUE)
normData200.lineplot <- normData200.lineplot[!(normData200.lineplot$inv %in% c("HsInv0124", "HsInv1110")), ]
lineplot<-ggplot(normData200.lineplot)+geom_line(aes(x = winPos, y = Means.Fold.Change, group = inv), color = "steelblue4")+
         facet_grid(Group ~ . ) +
         # geom_text(data = normData200.lineplot[normData200.lineplot$winPos == "inv_1", ] , aes(x = winPos, y = Means.Fold.Change, label = inv) )+
  geom_hline(yintercept = 1, color = "red", alpha = 0.75)

      
      # return(plot)

  plotname<- "report/2020-10-05_analysisNormalizedData/images/lineplot.200k.png"
  png(filename=plotname, width = 3508/5, height = 2480/3, units = "px" )
  plot<- annotate_figure(lineplot, top = text_grob(paste0("Log2 transformed fold changes between het. and homo in 3 groups" ), size = 10) )
  print(plot)
  dev.off()
  
  table(normData200.lineplot$Group)/11
  
  
  unique(normData200.lineplot$inv)
  
```

## 4: Fer plots nous

```{r}

# 200 KB
# 
# normData200 <- read.table("analysis/2020-10-14_09_crossovers_200k/crossoverResult_QN.txt", header = TRUE)

# Segona versió dels plots de línies, per grups, fent servir mitjanes de les finestres interiors de les invs i fold changes
normData200.boxplot <- plotInvsDifQN(normData200, genotypes=genotypes)

ggplot(normData200.boxplot)+geom_boxplot(aes(x = size ,y = Means.Fold.Change, fill = Group, group = inv ) ) +
         facet_grid(side ~ Group , scales = "free") + scale_x_log10()
         geom_text(data = normData200.lineplot[normData200.lineplot$winPos == "inv_1", ] , aes(x = winPos, y = Means.Fold.Changelog, label = inv)  )



  plotname<- "report/2020-10-05_analysisNormalizedData/images/lineplot.200k.png"
  png(filename=plotname, width = 3508, height = 2480, units = "px" )
  plot<- annotate_figure(lineplot, top = text_grob(paste0("Log2 transformed fold changes between het. and homo in 3 groups" ), size = 50) )
  print(plot)
  dev.off()
  
  
```


























