---
title: "Imputation report"
output:
  pdf_document: default
fig_caption: yes
---
```{r libraries, echo = FALSE, warning=FALSE, message=FALSE}
library(stringr)
library(ggplot2)
# library(ggpubr)
# library(naniar)
library(formattable)
library(kableExtra)
library(knitr)
# library(reshape2)
library("htmltools")
library("webshot")
# 
options(stringsAsFactors = FALSE)

# export_formattable <- function(f, file, width = "100%", height = NULL,
#                                background = "white", delay = 0.2)
#     {
#       w <- as.htmlwidget(f, width = width, height = height)
#       path <- html_print(w, background = background, viewer = NULL)
#       url <- paste0("file:///", gsub("\\\\", "/", normalizePath(path)))
#       webshot(url,
#               file = file,
#               selector = ".formattable_widget",
#               delay = delay)
#     }
```

# Context

Our goal is to quantify the effect of inversions in meiotic recombination, both in heterozygosis and homozygosis. We hypothesize that inverted regions may fail to synapse in heterozygous individuals and that homozygous individuals may have different distribution of crossovers depending on the orientation they carry. We have data from Bell et al. 2019, which sequence the genomes of 31,228 gametes from 20 sperm donors, identifying 813,122 crossover events. The first step in this project is to infer these individuals' genotypes of as many inversions as possible.

# Methods

Inversion genotypes were imputed with IMPUTE2, a genotype imputation and haplotype phasing program. Using a set of reference haplotypes genotyped for the inversions of interest and a genetic map, it infers the inversion status for our set of study individuals.

## Study panel preparation
We were provided with the VCFs for the 20 individuals in Bell et al. (2019) study. They were sequenced with a method called Sperm-seq: many individual sperm cells (974-2,274 gametes per donor) are sequenced to low coverage (median of ~1% of the haploid genome for each cell) simultaneously. Then, allelic haplotypes for the full length of every chromosome were inferred with ~40x coverage per donor.

The population of origin for each individual was provided by the sperm bank, but there were cases with conflicting ancestry information so I checked their ancestry with a Principal Component Analysis. The individuals from the 1000 Genomes Project and our study individuals were clustered using chromosome 1 with `qctools pca`. Then, our study individuals were classified according to the 1000 Genomes Project population cluster they fit into. **Figure 1** shows an example of the PCA result with some individuals. The final ancestry classifications are in **Table 1**. 



```{r, echo = FALSE, fig.cap="Principal Component Analysis for individuals NC6ab, NC22ab, NC25ab and NC26ab. Individuals NC6ab, NC22ab and NC25ab were reported by the sperm bank to be African American, while individual NC26ab was reported to be African American Asian. According to this analysis, individuals NC6ab and NC22ab were classified as AFR and individuals NC25ab and NC26ab as admixed.", fig.height=3.5}

load("Figure.Rdata")
pl

```

```{r, echo = FALSE}
PCA<-data.frame("Individual"=c("nc1","nc2","nc3","nc4","NC6ab","nc8","nc9","nc10","nc11","nc12","nc13","nc14","nc15","nc16","nc17","nc18","NC22ab","NC25ab","NC26ab","nc27"), "Reported.in.paper"= c(rep("European",4), "African American", rep("Asian", 3), rep("European", 8), "African American", "African American", "African American Asian", "Asian (conflicting ancestry information)"), "PCA.result"= c(rep("EUR",4), "AFR", rep("EAS", 3), rep("EUR", 8), "AFR", "Admixed", "Admixed", "SAS"))

# formattable(PCA, align="c", col.names=c("Individual", "Reported in original paper", "PCA result"))

kable(PCA, format = "latex", booktabs = T, caption = "Principal Component Analysis results.", col.names=c("Individual", "Reported in original paper", "PCA result"), align = "c") 
```

A main condition in IMPUTE2 is that variants must be mapped against the same genome build in the reference and test panels. In addition, it is important that all shared SNPs between reference and test panels are aligned to the same allele coding. Our reference panel is in hg39 and all variants are aligned to the '+' strand, while the provided VCFs were in hg38 and the strand alignment changed in some regions. To be sure that shared variants had exactly the same coordinates and strand alignment in both panels, I copied the 1000 Genomes Project Phase 3 coordinates into the test VCFs using the variant ID as a reference and created a file showing their strand orientation, which can be used with the IMPUTE2 option `-strand_g` to account for strand alignment changes along the sequence. 

## Reference panel preparation

Experimental genotypes for 111 inversions were available, 92 of which were autosomic. For each imputed inversion, a reference panel was created by merging these genotypes with the variants from the 1000 Genomes Project Phase 3 at 500 kb at each side of the inversion for the individuals in common.

Due to local differences in inversion recurrence I was advised to use as a reference panel only those individuals with the same ancestry as the individuals in the test panel. However, some of my individuals were admixed. With them, an alternative methodology can be used in which the reference panel icludes all the available individuals regardless of their ancestry, but IMPUTE2 uses only the closest n individuals to the test panel. To test which is the most suitable threshold, I imputed all the inversions in all the individuals in four ways: using all individuals from the same population as those in test panel (the first methodology, which can be considered a control) and using all individuals regardless of their ancestry but making IMPUTE2 to select the 500, 250 or 100 closest individuals to those in test panel (the second methodology with relaxed, medium and strict conditions, respectively).

### Inversions with deletion allele

Inversions HsInv0052 and HsInv0370 have three alleles: standard (STD), inverted (INV) and a deletion (DEL) that spans the inversion region. In the IMPUTE2 documentation, they explain that two or more variants can have the same position, for example a SNP and an INDEL. Following this example, I have divided these inversions into two variants. HsInv0052si and HsInv0370si refer to the standard and inverted alleles and HsInv0052nd and HsInv0370nd refer to not-deleted and deleted alleles. The genotype files were modified accordingly (**Table 2**).


```{r, echo = FALSE, fig.align='center', fig.height = 1}

PCA<-data.frame(
                "or"=c("STD/STD", "STD/INV", "INV/INV", "STD/DEL", "INV/DEL", "DEL/DEL"), 
                "si"= c("STD", "HET", "INV", "", "", ""),
                "nd"= c("STD", "STD", "STD", "HET", "HET", "DEL") )

# f<-formattable(PCA, align="c", col.names=c("Original genotype", "HsInv0052si and HsInv0370si", "HsInv0052nd and HsInv0370nd"))
# 
# as.htmlwidget(f)

kable(PCA, format = "latex", booktabs = T, caption = "Genotype file modifications", col.names=c("Original genotype", "HsInv0052si and HsInv0370si", "HsInv0052nd and HsInv0370nd"), align = "c") 
```

# Results

The output of IMPUTE2 is the probability for each genotype to be true. For each inversion and individual, the result with the highest probability has been included in a table and processed with R. However, only genotypes with a probability higher than 80% can be considered reliable. 

The results of imputation by individual and inversion are in the attached spreadsheet (sheet "Imputation results"). For each individual, inversions are tagged and sorted by imputability (with the categories "No_Polymorphic", "Tagged", "Imputable", "No_Imputable"). In addition, two columns summarize the reliability of the result by comparing the four methodological conditions that were tested:

* Uniformity: "good" when all four conditions report the same genotype, "alert" when all the conditions fot the second methodology predict the same genotype, which is different from that predicted by the first methodology and "bad" when conditions within the second methodology do not agree. Admixed individuals can only have "good" and "bad" Uniformity tags.
* Quality: "good" when all four conditions report probabilities >80% for their predicted genotypes, "alert" when some of the conditions report probabilities <80%, and "bad" when all conditions report <80% probabilites. 

For HsInv0052 and HsInv0370, the results for both parts of the imputation were interpreted (column Notes). In short, the "nd" part is used to know how many DEL alleles are there and the "si" part to know the content of the not-deleted alleles, i.e "nd" = HET + "si" = STD is interpreted as STD/DEL. In the case of DEL/DEL genotype, I expect "nd" to be  DEL with a good Uniformity and Quality and "si" to have not so good Uniformity or Qualities, unless the DEL haplotype is similar enough to STD or INV haplotypes.

Finally, the Uniformity and Quality of the imputation results was summarized by inversion and population and can be found in the attached spreadsheet (sheet "Summary by inversion"). 


```{r, echo = FALSE, include=FALSE, eval=FALSE}

# RESULTS OF IMPUTATION BY INDIVIDUAL AND INVERSION

setwd("~/20200227_RealRecombination/") 
populations<-  read.table("data/avery_individuals/samples_population.txt", header = TRUE)[,c(1,3)]
populations$sample<-str_split_fixed(populations$sample, "[.]", 2)[,2]

parse_vcf<-function(vcf){
  vcf<-vcf[,c(3,10:length(vcf))]
  
  vcf_melted<-reshape2::melt(vcf, id.vars = "Inversion")
  
  vcf_melted$Inversion<-str_split_fixed(vcf_melted$Inversion, ";", 2)[,1]
  vcf_melted<-cbind(vcf_melted, str_split_fixed(vcf_melted$value, ":", 2))
  
  vcf_merged<-merge(vcf_melted, populations , by.x = "variable", by.y = "sample", all.x = TRUE)

  colnames(vcf_merged)<-c("Individual", "Inversion", "Value", "Genotype", "Probability", "Population")
  
  vcf_final<-vcf_merged[,c("Inversion", "Individual", "Population", "Genotype", "Probability")]
  
  return(vcf_final)
}

process_directory<-function(directory){
  setwd(directory)
# SAMPLE
names<-c("Chromosome", "Position", "Inversion", "ref", "alt", "sc1", "sc2", "sc3", "format", populations$sample, "none" )
all_100<-  read.table( "100_ALL/100_ALL_re.vcf", sep = "\t", stringsAsFactors = FALSE, col.names = names )[,1:29]
all_250<-  read.table( "250_ALL/250_ALL_re.vcf", sep = "\t", stringsAsFactors = FALSE, col.names = names )[,1:29]
all_500<-  read.table( "500_ALL/500_ALL_re.vcf", sep = "\t", stringsAsFactors = FALSE, col.names = names )[,1:29]

# CONTROL
names_afr<-c("Chromosome", "Position", "Inversion", "ref", "alt", "sc1", "sc2", "sc3", "format", populations[populations$super_pop == "AFR","sample"], "none" )
names_eas<-c("Chromosome", "Position", "Inversion", "ref", "alt", "sc1", "sc2", "sc3", "format", populations[populations$super_pop == "EAS","sample"], "none" )
names_eur<-c("Chromosome", "Position", "Inversion", "ref", "alt", "sc1", "sc2", "sc3", "format", populations[populations$super_pop == "EUR","sample"], "none" )
names_sas<-c("Chromosome", "Position", "Inversion", "ref", "alt", "sc1", "sc2", "sc3", "format", populations[populations$super_pop == "SAS","sample"], "none" )
afr_con<-  read.table( "500_AFR/500_AFR_re.vcf", sep = "\t", stringsAsFactors = FALSE, col.names = names_afr )
eas_con<-  read.table( "500_EAS/500_EAS_re.vcf", sep = "\t", stringsAsFactors = FALSE, col.names = names_eas )
eur_con<-  read.table( "500_EUR/500_EUR_re.vcf", sep = "\t", stringsAsFactors = FALSE, col.names = names_eur )
sas_con<-  read.table( "500_SAS/500_SAS_re.vcf", sep = "\t", stringsAsFactors = FALSE, col.names = names_sas )

# rm(names, names_afr, names_eas, names_eur, names_sas)
# SAMPLE
all_100_p<-parse_vcf(all_100)
all_100_p$RefInd<-rep(100)
all_250_p<-parse_vcf(all_250)
all_250_p$RefInd<-rep(250)
all_500_p<-parse_vcf(all_500)
all_500_p$RefInd<-rep(500)

# CONTROL
afr_eas<-merge(afr_con[,-length(afr_con)], eas_con[,-length(eas_con)], by=c("Chromosome", "Position", "Inversion", "ref", "alt", "sc1", "sc2", "sc3", "format"), all = TRUE)
afr_eas_eur<-merge(afr_eas, eur_con[,-length(eur_con)], by=c("Chromosome", "Position", "Inversion", "ref", "alt", "sc1", "sc2", "sc3", "format"), all = TRUE)
afr_eas_eur_sas<-merge(afr_eas_eur, sas_con[,-length(sas_con)], by=c("Chromosome", "Position", "Inversion", "ref", "alt", "sc1", "sc2", "sc3", "format"), all = TRUE)

all_con_p<-parse_vcf(afr_eas_eur_sas)
all_con_p$RefInd<-rep("control")

# JOIN
# all_3r<-rbind(rbind(rbind(all_100_p, all_250_p), all_500_p), all_con_p)

colnames(all_100_p)[4:5]<-c("Genotype_100", "Probability_100")
colnames(all_250_p)[4:5]<-c("Genotype_250", "Probability_250")
colnames(all_500_p)[4:5]<-c("Genotype_500", "Probability_500")
colnames(all_con_p)[4:5]<-c("Genotype_con", "Probability_con")

by_cols<-c("Inversion", "Individual", "Population")
all_3c<-merge(merge(merge(all_100_p[,1:5], all_250_p[,1:5], by = by_cols), all_500_p[,1:5],  by = by_cols), all_con_p[1:5], by = by_cols, all.x=TRUE)
setwd("~/20200227_RealRecombination/")
return(all_3c)
# CLEAN
# rm(all_100, all_100_p, all_250, all_250_p, all_500, all_500_p,afr_eas,afr_eas_eur, afr_eas_eur_sas, afr_con, eas_con, sas_con, eur_con, all_con_p) # populations? 

}

directory<-"results/imputation/benchmarking_all/"
all_3c<-process_directory(directory)

# Replace "" by NA
indivs_vector<-all_3c$Individual
all_3c<- all_3c %>% replace_with_na_all(condition = ~.x == "")
all_3c<-as.data.frame(all_3c)
all_3c$Individual<-indivs_vector

#Sort by inversion imputability
imputability<-read.table("data/inversions_info/Inversions_imputability.txt", header = TRUE)
imputability$GLB<-factor(imputability$GLB, levels = c("No_Polymorphic", "Tagged", "Imputable", "No_Imputable"))
imputability$EUR<-factor(imputability$EUR, levels = c("No_Polymorphic", "Tagged", "Imputable", "No_Imputable"))
imputability$AFR<-factor(imputability$AFR, levels = c("No_Polymorphic", "Tagged", "Imputable", "No_Imputable"))

all_3c<-merge(all_3c,imputability, all.x = TRUE)

rm(imputability, indivs_vector)


all_3c$Conclusion<-""

# Known populations
all_3c[which(all_3c$Genotype_100 != all_3c$Genotype_250 | all_3c$Genotype_250!= all_3c$Genotype_500 | all_3c$Genotype_250!= all_3c$Genotype_100),"Conclusion"]<-"bad"
all_3c[which(all_3c$Genotype_100 == all_3c$Genotype_con & all_3c$Conclusion == "" ), "Conclusion"]<-"good"
all_3c[all_3c$Conclusion == "", "Conclusion"]<-"alert"
#Unknown populations
all_3c[(all_3c$Genotype_100 != all_3c$Genotype_250 | all_3c$Genotype_250!= all_3c$Genotype_500 | all_3c$Genotype_250!= all_3c$Genotype_100) & all_3c$Population == "ALL","Conclusion"]<-"bad"
all_3c[ all_3c$Conclusion != "bad" & all_3c$Population == "ALL", "Conclusion"]<-"good"


all_3c$Quality<-""

#Known populations
all_3c[which(all_3c$Probability_100 > 0.8 & all_3c$Probability_250 > 0.8 & all_3c$Probability_500 > 0.8 & all_3c$Probability_con > 0.8 ),"Quality"]<-"good"
all_3c[which(all_3c$Probability_100 < 0.8 & all_3c$Probability_250 < 0.8 & all_3c$Probability_500 < 0.8 & all_3c$Probability_con < 0.8 ),"Quality"]<-"bad"
all_3c[ all_3c$Quality == "","Quality"]<-"alert"
#Unknown populations
all_3c[(all_3c$Probability_100 > 0.8 & all_3c$Probability_250 > 0.8 & all_3c$Probability_500 > 0.8 ) & all_3c$Population == "ALL","Quality"]<-"good"
all_3c[(all_3c$Probability_100 < 0.8 & all_3c$Probability_250 < 0.8 & all_3c$Probability_500 < 0.8 )& all_3c$Population == "ALL","Quality"]<-"bad"
all_3c[ !(all_3c$Quality %in% c("good", "bad")) & all_3c$Population == "ALL","Quality"]<-"alert"



write.csv(all_3c[order(all_3c$GLB, all_3c$Individual), ], "individuals_imputation.csv")


```


```{r, echo = FALSE, include=FALSE, eval=FALSE}

# RESULTS OF IMPUTATION SUMMARIZED BY INVERSION

# my.pops<-c("EUR", "EAS", "SAS", "AFR", "ALL")
# res_list <- vector(mode = "list", length = 5)
# names(res_list)<-my.pops
# 
# all_3c$Conclusion<-as.factor(all_3c$Conclusion)
# all_3c$Genotype_con<-as.factor(all_3c$Genotype_con)
# all_3c<-all_3c[order(all_3c$GLB),]
# 
# for (pop in my.pops) {
#   
#   if (pop %in% "ALL"){
#     cols<-6
#     colist<-c( "Good_Geno" , "P_100",  "P_250","P_500", "Summary")
#   } else{
#     cols<-7
#     colist<-c( "Good_Geno" , "P_pop", "P_100",  "P_250","P_500", "Summary")
#   } 
#   
#   invs<-unique(all_3c$Inversion)
#   tmp<-data.frame(matrix(vector(), nrow = length(invs), ncol=cols, dimnames = list(c(1:length(invs)), c("Inversion",colist))), stringsAsFactors = FALSE )
#     tmp$Inversion<-invs
# 
#     for (inv in tmp$Inversion) {
#         
#         tmp_2<-all_3c[all_3c$Inversion == inv & all_3c$Population == pop,]
#         if(pop %in% "ALL"){
#             content<-c(table(tmp_2$Conclusion)["good"]/sum(table(tmp_2$Conclusion)),
#                 mean(as.numeric(tmp_2$Probability_100)),
#                 mean(as.numeric(tmp_2$Probability_250)),
#                 mean(as.numeric(tmp_2$Probability_500)) ,
#                sum(table(tmp_2[tmp_2$Conclusion == "good","Genotype_100"]) > 0))
#         }else{
#            content<-c(table(tmp_2$Conclusion)["good"]/sum(table(tmp_2$Conclusion)),
#              mean(as.numeric(tmp_2$Probability_con)),
#                 mean(as.numeric(tmp_2$Probability_100)),
#                 mean(as.numeric(tmp_2$Probability_250)),
#                 mean(as.numeric(tmp_2$Probability_500)) ,
#                sum(table(tmp_2[tmp_2$Conclusion == "good","Genotype_con"]) > 0))
#           
#         }
#         
#         tmp[tmp$Inversion == inv, colist] <- content
#               
#     }
# 
#     res_list[[pop]]<-tmp
#     rm(tmp, tmp_2)
# }
# 
# inversion_summary<-do.call(rbind, res_list)
# inversion_summary<-inversion_summary[,c(1:7, 9:14, 16:21, 23:28, 30:34)]

# RESULTS OF IMPUTATION SUMMARIZED BY INVERSION AND POPULATION

my.pops<-c("AFR", "EAS", "EUR", "SAS" , "ALL")
res_list <- vector(mode = "list", length = 5)
names(res_list)<-my.pops

all_3c$Conclusion<-as.factor(all_3c$Conclusion)
all_3c$Genotype_con<-as.factor(all_3c$Genotype_con)
all_3c<-all_3c[order(all_3c$GLB),]

for (pop in my.pops) {
  
    cols<-10
    colist<-c("Population", "Imputability","Good_Geno" , "P_pop", "P_100",  "P_250","P_500", "Summary", "Heteros")

    invs<-unique(all_3c$Inversion)
  
    tmp<-data.frame(matrix(vector(), nrow = length(invs), ncol=cols, dimnames = list(c(1:length(invs)), c("Inversion",colist))), stringsAsFactors = FALSE )
    tmp$Inversion<-invs
    
    for (inv in tmp$Inversion) {
        
        tmp_2<-all_3c[all_3c$Inversion == inv & all_3c$Population == pop,]
        
        if(pop %in% "ALL"){
            imput<-as.character(tmp_2[1,"GLB"])
             popmean<-NA
        }else if(pop %in% c("EUR", "AFR")){
           imput<-as.character(tmp_2[1,pop])
           popmean<- mean(as.numeric(tmp_2$Probability_con))
          
        }else{
           imput<-NA
           popmean<- mean(as.numeric(tmp_2$Probability_con))
        }
        
        content<-c(
             pop,
             imput,
             table(tmp_2$Conclusion)["good"]/sum(table(tmp_2$Conclusion)),
              popmean,
                mean(as.numeric(tmp_2$Probability_100)),
                mean(as.numeric(tmp_2$Probability_250)),
                mean(as.numeric(tmp_2$Probability_500)) ,
             sum(table(tmp_2[tmp_2$Conclusion == "good","Genotype_100"]) > 0),
             table(tmp_2[tmp_2$Conclusion == "good","Genotype_100"])["HET"] 
             )
        
        tmp[tmp$Inversion == inv, colist] <- content
              
    }

    res_list[[pop]]<-tmp
    # rm(tmp, tmp_2, content)
}

inversion_summary<-do.call(rbind, res_list)
inversion_summary<-inversion_summary[order(inversion_summary$Inversion),]

inversion_summary[is.na(inversion_summary$Heteros),"Heteros"]<-0
write.csv(inversion_summary, "inversions.csv")

```

# Discussion

According to these preliminary results, and taking into account only genotypes that can be considered correctly predicted (good Uniformity and Quality), 74 inversions (~80%) have heterozygous individuals. In addition, for 33 inversions (~35%) heterozygous and homozygous individuals for both orientations can be compared. The sample size for heterozygous individuals for each inversion is maximum 13 (**Figure 2**).


```{r, echo = FALSE, fig.cap="Distribution of sample sizes for valid heterozygous genotypes. ~20% of inversions do not have any heterozygous individuals. Most of the inversions have 9 or less valid heterozygous samples.", fig.height = 4}

# # I want to know how many inversions have three good genotypes globally
# 
# perinv<-sapply(unique(all_3c$Inversion), FUN= function(x){
#   sum(table(all_3c[all_3c$Inversion == x & all_3c$Conclusion == "good" & all_3c$Quality == "good", "Genotype_100"])>0)
# } )
# perinv<-data.frame(inversion=names(perinv), diff=perinv)
# # 
# # table(perinv$diff)
# 
# 
# perinv2<-sapply(unique(inversion_summary$Inversion), FUN= function(x){
#   sum(as.numeric(inversion_summary[inversion_summary$Inversion == x, "Heteros"]))
# } )
# 
# perinv2<-data.frame(inversion=names(perinv2), het=perinv2)
# 
# perinv<-merge(perinv, perinv2, by = "inversion")
# 
# save(perinv, file="data.Rdata")
load("data.Rdata")

ggplot(perinv)+geom_bar(aes(x = het), fill = "darkcyan" )+labs(x="Amount of heterozygous individuals", y = "Amount of inversions", title="Sample sizes available for heterozygous individuals")
# tab<-data.frame(table(perinv$diff))
# formattable(tab, align="c", col.names=c("Amount of different genotypes", "Number of inversions") )
# sum(table(perinv$het)[c(2:13)])

```

In general, imputation results are considered valid when they have good Uniformity and Quality. There are several ways in which the amount of valid results can be increased:

* Less strict filters: results with good Uniformity and probability >80% in their corresponding population, with poor performance when imputed using all individuals but good performance in their corresponding population or those with probability <80% with their own population, but excellent performance when using all the individuals as reference could also be accepted, depending on the inversion imputability. 
* In many cases, there are not enough SAS individuals genotyped to be used as a reference panel for individual nc27. This one could be filtered and interpreted as if it was admixed. 
* Inversion HsInv0058 showed very poor performance, which was surprising considering it has tag SNPs. The problem was that none of the SNPs around the region in the test VCFs is a tag SNP for HsInv0058. This problem can be addressed using breakseq to genotype it. I'm in process of downloading part of the data available in dbGAP to check the format and how raw is the data, which will determine how time-consuming this solution would be. 

