---
title: "Genotype inference analysis"
author: "Ruth GÃ³mez Graciani"
output:
  bookdown::pdf_document2:
    toc: false
    latex_engine: xelatex
  header-includes:
  - \usepackage{xcolor}
  - \usepackage{framed}


mainfont: OpenSans
---
\renewcommand{\topfraction}{.85}
\renewcommand{\bottomfraction}{.7}
\renewcommand{\textfraction}{.15}
\renewcommand{\floatpagefraction}{.66}
\setcounter{topnumber}{3}
\setcounter{bottomnumber}{3}
\setcounter{totalnumber}{4}

_In this report, I analyze the differences in local recombination rates between heterozygous and homozygous individuals for an inversion._

```{r setup, include=FALSE}

require("ggpubr")
require("reshape2")
library("ggplot2")
require("data.table")
require("knitr")
require("stringr")
library("kableExtra")
require("gridExtra")
library("ggh4x")
library(viridis)
require(colorspace)

# Example: how to store figure
  # plotname<- "report/2020-11-25_GraphicalRepresentationReport/lineplot.200k.small.png"
  # png(filename=plotname, width = 3508/5, height = 2480/3, units = "px" )
  # plot<- annotate_figure(lineplot, top = text_grob(paste0("Log2 transformed fold changes between het. and homo in 3 groups" ), size = 10) )
  # print(plot)
  # dev.off()


```

# Origin of the data

## Genotypes

```{r genoData, include=FALSE }

# Read genotypes
genotypes <- read.table("../2020-10-28_genotypeFilteringReport/allgenotypes_classified.csv", header = TRUE, sep = ",", row.names = 1, stringsAsFactors = FALSE)

# Use tagSNP genotype if available
genotypes$Genotype<-ifelse(is.na(genotypes$TagSNP.genotype), genotypes$Imp.genotype, genotypes$TagSNP.genotype)

# Filter out and keep only those accepted genotypes
genotypes<-genotypes[genotypes$Result %in% c("A.Imputed","A.Tag.Imput","A.Tagged" ), c("Individual", "Inversion","Population", "Genotype")]

```

Genotypes were imputed in our 20 individuals using IMPUTE2, tagSNP inference or both. Genotypes' quality control report can be found in "report/2020-10-28_genotypeFilteringReport/filteringAnalysis.pdf". We obtained more than 3 high-quality genotypes coming from both homozygous and heterozygous individuals for `r length(unique(genotypes$Inversion))` inversions.

## Map

```{r recmapData, include=FALSE }

# Read quantile normalized data
normData <- read.table("../../analysis/2021-02-18_09_crossovers/crossoverResult_QN.txt", header = TRUE)

```

Recombination maps were calculated from recombination events in a probabilistic way. The genome is divided into windows, for which recombination rates are calculated following a probabilistic method: instead of just assuming that the crossover took place in the center of the recombination event, each event is ponderated depending on how much of it is overlapping with a window, and the sum is used to calculate cM/Mb values for each window. Then, recombination results are normalized using a quantile normalization in order to make them comparable.

The effectivity of this method, as well as the smallest informative window size, were assessed with simulations. For each recombination event, a hypothetical actual location for the crossover was randomly selected and then the corresponding recombination rate calculated. We obtained the correlation between the simulated rates and the rates calculated with low-resolution recombination events. This gives us a measurement of how close estimated rates would be to real ones. The probabilistic method proved to be better than the center-point method, and window sizes between 150 and 200kb (corresponding to 0.9 and 0.95 correlations) would be optimal (Figure \@ref(fig:PLOTminimalWindow)).

```{r PLOTminimalWindow, echo = FALSE, message = FALSE, fig.cap ="Real recombination rates were simulated at different window sizes and comapred with the corresponding estimated ones. According to this result, our probabilistic method is more accurate than the center-point method, and the minimum informative window size is 150-200 kb.", fig.height=1, fig.pos="!H"}

include_graphics("..//2020-11-09_minimalInformativeWindow/window_assimptote.png", dpi = NA)

```

```{r joinGenoRecmap, include=FALSE }

# SOLVE NAME CHANGES

# Manually search for name changes between files
# g<-unique(genotypes$Inversion)
# n<- unique(normData$inv)
# g[!g%in%intersect(g,n)]

namechanges<-read.table("../../data/use/inversions_info/2021.01.19to38nameChanges", header = TRUE, stringsAsFactors = FALSE)
rownames(namechanges)<-namechanges$hg19
genotypes[genotypes$Inversion %in% namechanges$hg19, "Inversion"]<-namechanges[genotypes[genotypes$Inversion %in% namechanges$hg19, "Inversion"],"hg38"]

# MERGE  DATA
normGenoData<-merge(genotypes,normData, by.x =c("Individual", "Inversion" ), by.y =  c("ind", "inv")  )

# EXCLUDE DATA IF NEEDED
# normGenoData<-normGenoData[!(normGenoData$Inversion %in% c("HsInv0052id" ,"HsInv0052sd" ,"HsInv0052si" )),]

save(normGenoData, file = "normGenoData.RData")

```

```{r invinfoAndGroups, include=FALSE }

    invinfo<-read.table("../../data/use/inversions_info/2020-08_inversionInfo_hg38.csv", sep = "\t", header = TRUE, stringsAsFactors = FALSE)
  invsizes<-read.table("../../data/use/inversions_info/2020.07.inversions_hg38.csv", sep = "\t", header = FALSE, stringsAsFactors = FALSE)

  invinfo$Size<-NULL
  invsizes$Size<-invsizes$V6-invsizes$V3-1
  invsizes$SizeIn<-invsizes$V5-invsizes$V4-1

  invinfo<-merge(invinfo, invsizes, by.x = "Inversion", by.y = "V2", all = TRUE)
  invinfo<-invinfo[,c("Inversion", "Origin", "Size", "SizeIn")]

  # Now I set the groups according to the original distribution of sizes

  sizeThreshold<-quantile(invsizes$Size, 0.75)
  invinfo$SizeGroup<-ifelse(invinfo$Size > sizeThreshold, "Big", "Small")

```

```{r tidyDataFun, include =  FALSE }

# FUNCTION TO MAKE TIDY DATA
# Input: normGenoData table, inversions info table with anything relevant. At least the size!
# Options:
  # aggregateWins =  do I want to analyze inside-inversion windows separated or aggregated? separated: by-inversion detail plots; aggregated: summaries
  # aggregateInvs =  do I want to have one line per inversion with HET and HOMO means (aggregated), or one result per individual-inversion pair (separated)?


tidyData<-function(data=normGenoData, info = invinfo, aggregateWins , aggregateInvs){

  # Generate clean HET/HOMO tag
    normGenoData$Haplotype<-ifelse(normGenoData$Genotype %in% c("INV", "STD"), "Homozygous" , "Heterozygous")


  # aggregateWins

    if (aggregateWins == TRUE){
      # Windows aggregated by mean, with position from start to end of all the region

      # Value
      partn<-aggregate(cM.Mb.QN  ~  Inversion + Haplotype +position+Individual ,  normGenoData[normGenoData$position == "inv",], mean)
      # Start position
      parts<-aggregate(startWin  ~  Inversion + Haplotype +position+Individual ,  normGenoData[normGenoData$position == "inv",], min)
      # End position
      parte<-aggregate(endWin  ~  Inversion + Haplotype +position+Individual ,  normGenoData[normGenoData$position == "inv",], max)
      # Merge all 3 tables
      part<-merge(partn, merge(parts, parte))
      # Add window ID
      part$idWin<-paste(part$Inversion, "inv_1", sep="_" )

      # Join aggregated inversions with colindant windows
      toPlot<-rbind(normGenoData[normGenoData$position != "inv", colnames(part)], part)

    }else{
      # Use info as-is
      toPlot<-normGenoData
    }

  # aggregateInvs

    if (aggregateInvs == TRUE){
      # Aggregate by zygosity to make the two sides of the comparison
        # Mean value
        test.subset<-aggregate(   cM.Mb.QN  ~  Inversion + startWin+endWin + Haplotype +idWin,toPlot, mean)
        # Count of individuals
        toPlot$counts<-1
        test.counts<-aggregate(   counts  ~  Inversion + startWin+endWin + Haplotype +idWin,toPlot, sum)
        test.subset<-merge(test.subset, test.counts)

      # Make both tables wide
        # Counts table wide
        test.subset.counts<-reshape2::dcast(test.subset, Inversion + startWin + endWin + idWin  ~ Haplotype, value.var="counts")
        colnames(test.subset.counts)[colnames(test.subset.counts) %in% c( "Heterozygous", "Homozygous"  )]<-
          c("Heterozygous.counts", "Homozygous.counts"  )
        # CM.Mb table wide
        test.subset.values<-reshape2::dcast(test.subset, Inversion + startWin + endWin + idWin  ~ Haplotype, value.var="cM.Mb.QN")
        colnames(test.subset.values)[colnames(test.subset.values) %in% c( "Heterozygous", "Homozygous"  )]<-
          c("Heterozygous.cM.Mb.QN", "Homozygous.cM.Mb.QN" )

      # Merge wide table
        test.subset.wide<-merge(test.subset.counts, test.subset.values)

      # Make comparisons
        # Difference
        test.subset.wide$Means.Difference<- test.subset.wide$Heterozygous.cM.Mb.QN - test.subset.wide$Homozygous.cM.Mb.QN
        # Fold changes
        test.subset.wide$Means.Fold.Change<- test.subset.wide$Heterozygous.cM.Mb.QN / test.subset.wide$Homozygous.cM.Mb.QN
        # Fold changes log
        test.subset.wide$Means.Fold.Changelog<- log2(test.subset.wide$Means.Fold.Change)


        
      # Discard those without data
        test.subset.wide<-  test.subset.wide[!is.na(test.subset.wide$Means.Difference),]
    }else{
      # Use info as-is
      test.subset.wide<-toPlot
      
    }


  # Order table and factors

      # Relative positions and window IDs
      test.subset.wide[c("side", "relpos")]<-str_split_fixed(test.subset.wide$idWin, "_", 3)[,c(2,3)]
      test.subset.wide[c("winPos")]<-str_split_fixed(test.subset.wide$idWin, "_", 2)[,2]

      test.subset.wide$side<-factor(test.subset.wide$side, levels = c("left", "inv", "right"))
      test.subset.wide$relpos<-as.numeric(test.subset.wide$relpos)

      test.subset.wide<-test.subset.wide[order( test.subset.wide$side, test.subset.wide$relpos),]
      test.subset.wide$winPos<- factor(  test.subset.wide$winPos, levels = unique(  test.subset.wide$winPos))


  # Add relevant inversion info

      test.subset.wide<-merge(test.subset.wide, invinfo, all.x = TRUE)

      return(test.subset.wide)

}

# testFF<-tidyData(aggregateWins = F, aggregateInvs = F)
testFT<-tidyData(aggregateWins = FALSE, aggregateInvs = T) # >1 window per inversion, 1 row per inversion window
# testTF<-tidyData(aggregateWins =T, aggregateInvs = FALSE) # 1 window per inversion, 1 row per inversion-window-individual
testTT<-tidyData(aggregateWins = T, aggregateInvs = T) # 1 window per inversion,1 row per inversion window

```

# Figures

> Include only the most relevant figures
> 
> -- For later


-------------------------------------------

# Appendix: all the tested figures {-}

_This appendix includes all the tested figures to avoid having to repeat them or not remembering if we tested something._

# Data selection
 
Here I list all the variables and data trimming criteria that were at some point considered or suggested. 

* **Nominal variables:**  
    - Breakpoint type (simple/complex)
    - Presence of inverted repeats
    - Breakpoint definition (Very well characterized and reliable vs. Not sure)
    - Physical length (natural or transformed to a normal distribution)
    - Genetic length (natural or transformed to a normal distribution)
* **Measurement variables :** 1 value per inversion - and window, if analyzing more than one window.
    - Fold change of the Normalized Recombination Rate between heterozygous and homozygous individual means
    - P-value of the comparison between heterozygous and homozygous means through a Student's T test
    - Power of the aforementioned Student's T test
* **Data trimming criteria:** aka, data to remove.
    - Outliers
    - Inversions <1kb

    
    
# Raw measurements visualization

## Size bias in genotyped inversions. 

Size distributions (log transformed) for all the available inversions (expected) and for the actually genotyped inversions. Small inversions are genotyped proportionally to the original distribution while big inversions are less often correctly genotyped, probably because they tend to be NAHR-mediated, recurrent inversions. 

```{r PLOTgenotypingBias, echo = FALSE, fig.cap ="", fig.pos="!H"}

expected <- data.frame(size =invinfo$Size ,logsize = log10(invinfo$Size), source = "expected")
genotyped<-data.frame(size=unique(testTT[,c("Size", "Inversion")])$Size, logsize=(log10(unique(testTT[,c("Size", "Inversion")])$Size)), source = "genotyped")

sizeComp<-rbind(expected, genotyped)


ggplot(sizeComp, aes(x = logsize))+geom_histogram(bins = 10)+facet_wrap(.~source , scales = "free")

```


## Crossover events overlapping with the studied region around genotyped inversions

```{r FUNregionRawVisualization, echo = FALSE, fig.cap ="", fig.pos="!H",  warning=F, message=F,  fig.height=6}

crossovers_raw<-read.csv("../../data/use/avery_crossovers/allcrossovers.bed", sep = "\t")

viewRaw<-function(inversion){

coordinates<-invsizes[invsizes$V2 == inversion, c("V1", "V3", "V6")]
colnames(coordinates)<-c("chr", "start", "end")

windows<-normGenoData[normGenoData$Inversion == inversion, ]

crossovers_raw_subset<-crossovers_raw[crossovers_raw$chr == as.character(unique(windows$chrWin))&
                                        (
                                          crossovers_raw$pos.leftbound < max(windows$endWin) & 
                                            crossovers_raw$pos.rightbound> min(windows$startWin)
                                        ), ]

crossovers_raw_subset$eventID<-rownames(crossovers_raw_subset)
crossovers_raw_subset<-reshape2::melt(crossovers_raw_subset, id.vars = c( "chr", "donor", "eventID" ))

crossovers_raw_subset<-merge(crossovers_raw_subset, unique(windows[,c("Individual", "Genotype", "samples")]), by.x = "donor", by.y = "Individual")



ggplot(crossovers_raw_subset ) + 
  geom_line(aes(x = value, y = eventID, color = samples))+
  geom_text(data = unique(crossovers_raw_subset[,c("donor", "samples", "Genotype")]), aes(x = -Inf, y = Inf, label = samples, color = samples), hjust = -0.2, vjust = 1.5 )+
  facet_nested(Genotype+donor~., scales = "free")+
  geom_vline(aes(xintercept = coordinates$start))+
  geom_vline(aes(xintercept = coordinates$end))+
  geom_vline(aes(xintercept =min(windows$startWin)))+
  geom_vline(aes(xintercept =  max(windows$endWin)))+
  scale_color_continuous_sequential(palette = "Viridis", begin = 0.05, end =1) +
  ggtitle( paste0("Crossover events around ", inversion) )+
  theme( 
    axis.title.y=element_blank(),axis.text.y=element_blank(),axis.ticks.y=element_blank(),
                  legend.text = element_text(size = 7),
          legend.key.size = unit(0.5, "line"),
          legend.margin =margin(t=0.5,b=0.5), 
        legend.title = element_text(size = 8))
  
}

viewRaw("HsInv0325")

```

\pagebreak

## Quantile-Normalized Recombination Rate values distribution in genotyped inversions

```{r FUNlnDist, echo = FALSE, fig.cap ="", fig.pos="!H"}

# Make function to return plot
# I only need the column cM.Mb.QN to be present
PLOTlognormalDistriutions<-function(normGenoData){
  # Make delta-lognormal adapted column 
  replacementValue<- min(normGenoData[normGenoData$cM.Mb.QN>0,"cM.Mb.QN" ])/2
  normGenoData$cM.Mb.QN_dln<-ifelse(normGenoData$cM.Mb.QN == 0, replacementValue, normGenoData$cM.Mb.QN)

  # General distribution
  plot1<-ggplot(normGenoData)+geom_histogram(aes(x = cM.Mb.QN))
  
  # Delta.lognormal distribution
  zeroData<-normGenoData[normGenoData$cM.Mb.QN_dln == min(normGenoData$cM.Mb.QN_dln),]
  
  highlightColor<-"#008080"

  plot2<-ggplot()+
        geom_histogram(data = normGenoData, aes(x = log10(cM.Mb.QN_dln)))+
        geom_histogram(data = zeroData,aes(x = log10(cM.Mb.QN_dln)), fill = highlightColor)+
        geom_text(data = , aes(x = log10(replacementValue), y = nrow(zeroData), label = paste0("Amount of 0 values: ", nrow(zeroData)) ),
                  vjust = -1, hjust = 0.1, color =highlightColor )
  
  return(list(plot1, plot2)) 
  
}
  

```

### General distribution 

All windows and inversions.

```{r PLOT.General.lnDist, echo = FALSE, fig.cap ="", fig.pos="!H", warning=F, message=F,fig.height=3.5}
plots<-PLOTlognormalDistriutions(normGenoData)

ggarrange( plotlist = plots, nrow = 1)
```

### Inside inversion distribution 

All windows from inside inversions (some inversions have more windows than others).

```{r PLOT.Inside.lnDist, echo = FALSE, fig.cap ="", fig.pos="!H", warning=F, message=F,fig.height=3.5}

plots<-PLOTlognormalDistriutions(normGenoData[normGenoData$position == "inv" ,])


ggarrange( plotlist = plots, nrow = 1)

```

### Mean inside inversion windows distribution 

The mean value for the windows inside inversions (one window per iversion and individual).

**To make this plot better, make scales for Heterozygous and homozygous equal! Note that heterozygous individuals have lower values in general**

```{r PLOT.InsideMeans.lnDist, echo = FALSE, fig.cap ="", fig.pos="!H", warning=F, message=F,fig.height=7}

moltenTestTT<-reshape2::melt( testTT[testTT$side == "inv",], id.vars = c("Inversion", "idWin"), measure.vars = c("Heterozygous.cM.Mb.QN", "Homozygous.cM.Mb.QN") , variable.name = "Zygos", value.name = "cM.Mb.QN")

p<-PLOTlognormalDistriutions(moltenTestTT[moltenTestTT$Zygos == "Heterozygous.cM.Mb.QN",])
q<-PLOTlognormalDistriutions(moltenTestTT[moltenTestTT$Zygos == "Homozygous.cM.Mb.QN",])

ggarrange(
  annotate_figure(ggarrange( plotlist=p, ncol = 1),  top = text_grob("Heterozygous") ),
  annotate_figure(ggarrange( plotlist=q, ncol = 1),  top = text_grob("Homozygous") )
)
  
```

\pagebreak

## Local recombination rate panoramic visualization in genotyped inversions

### All data

This is only a sample, but a figure with all the inversions or only a selected group can be generated

```{r PLOT.panoramicAll, echo = FALSE,  fig.pos="!H", warning=F, message=F, fig.height = 7}
    
panoramicPlot<-function(normGenoData){
  normGenoData$Genotype <- as.factor(normGenoData$Genotype)
  plotlist<-list()
  for(inversion in unique(normGenoData$Inversion)){
    
    test.subset<-normGenoData[normGenoData$Inversion == inversion,]
    
  plotlist[[inversion]]<-ggplot(test.subset)+
    geom_rect(aes(xmin=startWin, xmax=endWin, ymin = 0, ymax = cM.Mb.QN  )) +
    facet_wrap(Genotype ~    Individual, labeller = label_wrap_gen(multi_line=FALSE)) +
    geom_rect(data = test.subset[test.subset$position == "inv",], aes(xmin=startWin, xmax=endWin, ymin = 0, ymax = cM.Mb.QN, fill = Genotype))+
    scale_fill_manual(values = c("#edae49", "#66a182", "#00798c"), labels = c("HET", "INV","STD"), drop = FALSE)+
    ggtitle(inversion) +
    theme(axis.title.x=element_blank(),axis.text.x=element_blank(),axis.ticks.x=element_blank())
    
    
  }
  return(plotlist)  
}

 
plotlist<-panoramicPlot(normGenoData)

# Now different plots can be called
  # dir.create("report/2020-10-05_analysisNormalizedData/images")
  
  # printplot
  # plotname<- "report/2020-10-05_analysisNormalizedData/images/allinvs.b.200k.png"
  # png(filename=plotname, width = 3508, height = 2480, units = "px" )
  # plot<- 

invstoPlot<-tail(unique(testTT[order(testTT$Size),"Inversion"]),2)

    annotate_figure( ggarrange(plotlist=plotlist[invstoPlot], ncol = 1, common.legend = TRUE), top = text_grob(paste0("Quantile normalized recombination rates" )) )
  # print(plot)
  # dev.off()
```

\pagebreak


### Homozygous vs. Heterozygous means

This is only a sample, but a figure with all the inversions or only a selected group can be generated.

```{r PLOT.panoramicHetHom, echo = FALSE,  fig.pos="!H", warning=F, message=F, fig.height = 7}

    
panoramiHetHomPlot<-function(testTT){

  plotlist<-list()
  for(inversion in unique(testTT$Inversion)){

    test.subset<-testFT[testFT$Inversion == inversion,]
    test.subset<-reshape2::melt(test.subset, id.vars=c("startWin", "endWin", "Inversion", "side"),  measure.vars = c("Heterozygous.cM.Mb.QN", "Homozygous.cM.Mb.QN") , variable.name = "Zygosity",value.name =  "cM.Mb.QN")
    test.subset$Zygosity<-ifelse( test.subset$Zygosity == "Heterozygous.cM.Mb.QN", "Heterozygous", "Homozygous")
    
  plotlist[[inversion]]<-ggplot(test.subset)+
    geom_rect(aes(xmin=startWin, xmax=endWin, ymin = 0, ymax = cM.Mb.QN  )) +
    facet_wrap(Zygosity ~ ., labeller = label_wrap_gen(multi_line=FALSE)) +
    geom_rect(data = test.subset[test.subset$side == "inv",], aes(xmin=startWin, xmax=endWin, ymin = 0, ymax = cM.Mb.QN, fill = Zygosity))+
    scale_fill_manual(values = c("#edae49", "#66a182"))+
    ggtitle(inversion) +
    theme(axis.title.x=element_blank(),axis.text.x=element_blank(),axis.ticks.x=element_blank(), legend.position = "none")
    
    
  }
  return(plotlist)
}

 
plotlist<-panoramiHetHomPlot(testTT)

# Now different plots can be called
  # dir.create("report/2020-10-05_analysisNormalizedData/images")
  
  # printplot
  # plotname<- "report/2020-10-05_analysisNormalizedData/images/allinvs.b.200k.png"
  # png(filename=plotname, width = 3508, height = 2480, units = "px" )
  # plot<- 

invstoPlot<-tail(unique(testTT[order(testTT$Size),"Inversion"]),8)

    annotate_figure( ggarrange(plotlist=plotlist[invstoPlot], common.legend = TRUE, ncol = 2, nrow = 4), top = text_grob(paste0("Quantile normalized recombination rates (Zygosity means)" )) )
  # print(plot)
  # dev.off()

```

\pagebreak

# Heterozygous vs. Homozygous formal comparison

```{r FUNallPanel, echo = FALSE, fig.cap ="", fig.pos="!H"}

 # Hacer una pagina de plots, donde se incluyen todos los plots relevantes/conocidos que hemos estado mirando. 
 # Boxplot x tamaÃ±os, lineplot x tamaÃ±os, correlacion con tamaÃ±o fisico, log fisico?, genetico, log genetico?
# Maybe other kind of non-linear regression or robust regression

PLOTpanel<-function(testTT, title){

# c("#edae49", "#66a182", "#00798c")
  
# Line plot with all the region
  line1<-ggplot(testTT)+geom_line(aes(x = winPos, y = Means.Fold.Changelog, group = Inversion), color = "steelblue4")+
         facet_grid(SizeGroup ~ . ) +
  geom_hline(yintercept = 0, color = "red", alpha = 0.5)+
  ggtitle("Line Plot (all region)")+
    theme(plot.title = element_text( size=9))+
     theme(axis.title.x=element_blank(),axis.text.x=element_text(angle = 45, vjust = , hjust=1))

  # Line plot of center vs. extremes
  line2<-ggplot(testTT[testTT$winPos %in% c("left_1", "inv_1", "right_5") ,])+geom_line(aes(x = winPos, y = Means.Fold.Changelog, group = Inversion), color = "steelblue4")+facet_grid(SizeGroup ~ . ) +
  geom_hline(yintercept = 0, color = "red", alpha = 0.5)+
  ggtitle("Line Plot (inversion vs. extremes)")+
    theme(plot.title = element_text( size=9),axis.title.x=element_blank())
  
  # "Amount of cases with Heterozygous > Homozygous and viceversa
  # Remember! Het/Homo

  testTable<-testTT[testTT$side == "inv",]
  
  testTable$difgroup<-ifelse(testTable$Means.Fold.Changelog>0, "HET > HOM", "HET < HOM" )
  
  posneg<-ggplot(testTable)+geom_bar(aes(x = SizeGroup, fill = difgroup), position = "dodge")+
    ggtitle("Qualitative results \ncounts")+scale_fill_manual(values = c("#00798c", "#66a182"))+
    theme(legend.position = "top", legend.title = element_blank() , 
          legend.text = element_text(size = 5),
          legend.key.size = unit(0.5, "line"),
          legend.margin =margin(t=0.5,b=0.5)
          )+
    geom_text(stat='count', aes(label=..count.., x = SizeGroup, group = difgroup), position = position_dodge(width = .8), vjust = 2)+
    theme(plot.title = element_text( size=9),axis.title.x=element_blank(), axis.title.y=element_blank())
    
  
  # Boxplot in all the region
  boxplot<-ggplot(testTT, aes(y = Means.Fold.Changelog, x = winPos))+
    geom_boxplot(outlier.size =  -Inf )+
    geom_jitter( width = 0.1, alpha = 0.3)+
    facet_grid(SizeGroup ~ . )+
    geom_hline(aes(yintercept = 0), color = "red", alpha = 0.5)+
  ggtitle("Box Plot (all region)")+
    theme(plot.title = element_text( size=9),axis.title.x=element_blank())


  # Correlation with size
 correlsize<- ggplot(testTT[testTT$winPos == "inv_1", ], aes(x = Size, y = Means.Fold.Changelog)) + 
   geom_point() + stat_smooth(method = "lm", col = "#00798c") +  
   ggtitle("Size vs. Fold change in Het vs Hom")+ 
 stat_cor(method = "pearson", label.x.npc = 0, label.y.npc = 1 ) +
    theme(plot.title = element_text( size=9))
  
  # Correlation with logsize
      
   testTT$logSize<-log10(testTT$Size)
     correlogsize<-ggplot(testTT[testTT$winPos == "inv_1", ], aes(x = logSize, y = Means.Fold.Changelog)) + 
   geom_point() + stat_smooth(method = "lm", col = "#00798c") +  
   ggtitle("log10(Size) vs. Fold change in Het vs Hom")+ 
 stat_cor(method = "pearson", label.x.npc = 0, label.y.npc = 1 ) +
    theme(plot.title = element_text( size=9))
    
    return(  annotate_figure( ggarrange( ggarrange(plotlist = list(line1, line2, posneg), nrow = 1, widths = c(2,2,1.5)),
                                         boxplot, 
                                         ggarrange(correlsize, correlogsize, nrow = 1),  
                                         nrow = 3, heights = c(1, 1.5, 1) ), 
                              top = text_grob(title)
                            )   
          )
     
     
    
}
```

## All the dataset

```{r PLOTPanelAlldata, echo = FALSE, fig.cap ="",  fig.pos="!H", warning=F, message=F,  fig.height=8.5}

PLOTpanel(testTT,"All the dataset" )


```


## Dataset without outliers

```{r PLOTPanelNo325, echo = FALSE, fig.cap ="",  fig.pos="!H", warning=F, message=F,  fig.height=8.5}

PLOTpanel(testTT[!(testTT$Inversion %in% c("HsInv0325") ),],"HsInv0325 excluded" )


```

```{r PLOTPanelNo501, echo = FALSE, fig.cap ="", fig.pos="!H",  warning=F, message=F,  fig.height=9}

PLOTpanel(testTT[!(testTT$Inversion %in% c("HsInv0501") ),],"HsInv0501 excluded" )



```
```{r PLOTPanelNoOutliers, echo = FALSE, fig.cap ="", fig.pos="!H",  warning=F, message=F,  fig.height=9}


PLOTpanel(testTT[!(testTT$Inversion %in% c("HsInv0325", "HsInv0501") ),],"HsInv0325 and HsInv0501 excluded" )



```


```{r}
 # testTT$logSize<-log10(testTT$Size)origin


p<-   ggplot(testTT[testTT$winPos == "inv_1" & !(testTT$Inversion %in% c("HsInv0325", "HsInv1256")) , ], aes(x = Size, y = Means.Fold.Changelog)) + 
   geom_point(aes( color=Origin)) +
     stat_smooth(method = "lm", col = "#00798c") +  stat_smooth(method = "lm", aes(color=Origin, fill = Origin), alpha = 0) +  
   ggtitle("Inversion size effect in crossover rate decrease")+ 
 stat_cor(method = "pearson", label.x.npc =1, label.y.npc = 1, hjust=1, col = "#00798c" ) +
     stat_cor(method = "pearson", label.x.npc =1, label.y.npc = 0.95, hjust=1 , aes(color=Origin)) +
  scale_x_log10(breaks = scales::trans_breaks("log10", function(x) 10^x),   labels = scales::trans_format("log10", scales::math_format(10^.x)) )  +
          annotation_logticks(sides = "b")+
  xlab("Inversion size")+ ylab("log2 fold change of average crossover rates inside inversion  \n( Heterozygous / Homozygous ) ")
   
   
   ggsave( plot = p, filename = "Rplot_no325-1256-501.png")
   
```

```{r}
#CORRELATION TEST
library(MASS)
library(ggplot2)
library(scales)
testTT[testTT$Inversion == "HsInv0501", "Origin"]<-"NAHR"
testTT[testTT$Inversion == "HsInv0191", "Origin"]<-"NH"

posterPlot<-function(table, name){
  standardColor<-"#0d4b63"
 p<- ggplot(table, aes(x = Size, y = Means.Fold.Changelog)) + 
   # Points, colored
    geom_point(aes( color=Origin)) +
   # Statistics by group
    # stat_cor(method = "pearson", label.x.npc =1, label.y.npc = 0.95, hjust=1 , aes(color=Origin)) +
     # stat_smooth(method = "lm", aes(color=Origin, fill = Origin), alpha = 0) +  
   # Statistics global
    stat_smooth(method = "lm", col = standardColor) +  
    stat_cor(method = "pearson", label.x.npc =1, label.y.npc = 1, hjust=1, col = "black" ) +
   # Guide 0
    geom_hline(yintercept = 0, color =  "#c24f41", linetype="dashed", alpha = 1)+
   # Guide size
    geom_vline(xintercept = sizeThreshold, color = standardColor, linetype = "dashed", alpha = 0.5)+
    annotate(geom = "text", x = sizeThreshold, y = -3.8, label = paste0("3rd quartile in size = ", round(sizeThreshold/1000, 2), " kb"),
             hjust = -0.05, color = standardColor, alpha = 1)+
   # X axis log
    scale_x_log10(breaks = scales::trans_breaks("log10", function(x) 10^x),   labels = scales::trans_format("log10", scales::math_format(10^.x)) )  +
    annotation_logticks(sides = "b")+
    # Legend position
    theme(legend.position = c(0.2, 0.2))+
   # Outside labels
    xlab("Inversion size")+  
    ylab("log2 fold change of average crossover rates inside inversion  \n( Heterozygous / Homozygous ) ")+
    ggtitle("Inversion size effect in crossover rate decrease")


   return(p)
   ggsave( plot = p, filename = paste0("Rplot_","posterPlot",".png" ), width = 929, height = 767, units = "px", dpi = 300, scale = 2)
}
   

p<-posterPlot(testTT[testTT$winPos == "inv_1" & !(testTT$Inversion %in% c("HsInv0325")) , ], "no325")

posterPlot(testTT[testTT$winPos == "inv_1" & !(testTT$Inversion %in% c("HsInv0325", "HsInv1256", "HsInv0501")) , ], "no325-1256-501")
posterPlot(testTT[testTT$winPos == "inv_1" & !(testTT$Inversion %in% c("HsInv0325", "HsInv0501")) , ], "no325-501")
posterPlot(testTT[testTT$winPos == "inv_1" & !(testTT$Inversion %in% c("HsInv0501")) , ], "no501")

posterPlot(testTT[testTT$winPos == "inv_1"  , ], "allinvs")

#BINOMIAL TEST
binomTest<-function(table){
  
  samples<-nrow(table)
  success<-nrow(table[table$Means.Fold.Changelog > 0 , ])

   # H0: The coin is not biased. 
   # H1: The coin is biased in favour of FC<0 (tails)
   # We need to calculate more than just the probability that it lands on FC>0 (heads) "success" times. If it landed on FC>0 (heads) fewer than "success" times, that would be even more evidence that the coin is biased in favour of FC<0 (tails).Consequently we need to add up the probability of it landing on  FC>0 (heads) 1 time, 2 times, â¦ all the way up to "success" times. IF (P[Xâ¤"heads"] < 0.05) THEN H1 (tails win more). AKA IF (P[Xâ¤"FC>0 "] < 0.05) THEN H1 (FC<0  win more)
  
    if (pbinom(success, samples, 0.5) < 0.05){#cumulative probability P[Xâ¤"FC>0 "]
    sprintf("We had %i inversions recombining more in heterozygotes from %i samples, P[Xâ¤FC>0] is %f, which is smaller than 0.05, so the ratio is biased in favor of FC < 0.",success, samples, pbinom(success, samples, 0.5 )   )
  } else{
        sprintf("We had %i inversions recombining more in heterozygotes from %i samples, P[Xâ¤FC>0] is %f, which is larger than 0.05, so the ratio is not biased.",success, samples, pbinom(success, samples, 0.5 )   )
  }
  
}

binomTest(testTT[testTT$winPos == "inv_1", ])
binomTest(testTT[testTT$winPos == "inv_1"& testTT$SizeGroup == "Small" , ])
binomTest(testTT[testTT$winPos == "inv_1"& testTT$SizeGroup == "Big" , ])

   
```

> binomTest(testTT[testTT$winPos == "inv_1", ])
> [1] "We had 24 inversions recombining more in heterozygotes from 59 samples, P[Xâ¤FC>0] is 0.096263, which is larger than 0.05, so the ratio is not biased."
> binomTest(testTT[testTT$winPos == "inv_1"& testTT$SizeGroup == "Small" , ])
> [1] "We had 24 inversions recombining more in heterozygotes from 53 samples, P[Xâ¤FC>0] is 0.291566, which is larger than 0.05, so the ratio is not biased."
> binomTest(testTT[testTT$winPos == "inv_1"& testTT$SizeGroup == "Big" , ])
> [1] "We had 0 inversions recombining more in heterozygotes from 6 samples, P[Xâ¤FC>0] is 0.015625, which is smaller than 0.05, so the ratio is biased in favor of FC < 0."

# Modelo nulo con genotipos al azar

# HET vs HOMO in Student's T p-value. 1 point = 1 inv
* Boxplot per finestres - left in right, left i right haurien de ser mes iguals que in
* CorrelaciÃ³ mida fÃ­sica vs pval
* Correlacio mida genÃ¨tica vs pval

# HET vs HOMO in power calculations. 1 point = 1 inv

# Chi square - proporciÃ³ de positius i negatius

* Plot de positius i negatius
* Chi square de les proporcions entre positius i negatius, separats per mida

# Outliers


* Outlier selection
* Outlier explanation
* Solapament entre events de recombinatiÃ³ (raw data) i inversions. 




* Plot de panorama (normalized recombination rate) per 1 sola inversio. 
  * Individu a individu
  * Mitjana HET vs HOMO
* p-value of the normalized rec.rate in the out distribution, to compare between invs, especially for outliers?


