---
title: "Anàlisi estadístic"
output: html_notebook
editor_options: 
  chunk_output_type: console
---

```{r setLibraries}

setwd("../../")

library(ggplot2)

```


# Objectius i preguntes

L'objectiu general d'aquesta part és detectar diferències en el panorama de recombinació meiòtica en individus i/o regions que no presenten una inversió especte els que sí, controlant o tenint en compte diferents propietats intrínseques de les diferents inversions.

Les preguntes que podriem intentar contestar serien les següents:

* Les inversions tenen tendència a caure en zones d'alta o baixa recombinació? Aquesta pregunta ja la vam intentar contestar amb les dades de recombinació.
* Com afecta cada genotip de la inversió a les taxes de recombinació?
* Si comparem les taxes de recombinació històrica i real al voltant de la inversió, redueixen la recombinació històrica les inversions? En quina proporció?
* Està algun dels efectes mencionats relacionat amb propietats de les inversinons com mida o mètode de generació?

# Origen de les dades

Els individus de prova i els events de recombinació provenen de Bell et al. 2019. Les coordenades de les inversions de InvFEST v. hg38.

Tot i que la resolució dels events de recombinació no és la que esperàvem, la gran quantitat de cèl·lules analitzades per individu en alguns casos permet veure patrons en el nombre d'entrecreuaments que podrien estar tenint lloc a una resolució una mica millor de la que tindrem a priori (diapositiva 2, apartat Dades de recombinació).

# Anlisi descriptiu

```{r upload_files}

# This should be revised!! which merges do I want to be "all"??

# Crossover maps

  # Upload
  crossover_maps<-read.table("analysis/2020-05-29_09_crossovers/windows_x_crossovers_weighted.txt")
  colnames(crossover_maps)<-c("chr_w", "start_w", "end_w", "chr_e", "start_e", "end_e", "ind", "overlap", "score")

  # Summarize
  crossover_maps$ind<-sub("_.*$","", crossover_maps$ind)
  crossover_summary<-aggregate(score ~ chr_w+start_w+end_w+ind, data=crossover_maps, FUN=sum)
  
# Inversions overlapping with crossovers
  
  # Upload
  crossover_invbreak<-read.table("analysis/2020-05-29_09_crossovers/windows_x_inversions.txt", sep = "\t")
  colnames(crossover_invbreak)<-c("chr_i", "source", "type", "start_i", "end_i", "a", "b", "c", "info", "chr_w","start_w", "end_w" )
  
  # Extract inversions name
  crossover_inv<-crossover_invbreak[crossover_invbreak$type == "inversion",c("chr_i", "start_i", "end_i", "info", "chr_w","start_w", "end_w" )]
   crossover_inv$info<-sub("^ID=\\d+;Name=" , "", crossover_inv$info)
   crossover_inv$info<-sub(";.*$" , "", crossover_inv$info)
   
  # Get score for windows overlapping with inversions
   invscores<-merge(crossover_inv, crossover_summary)
   
# Control by number of samples  
  numofsamples<-read.table("report/2020-04-29_statisticalPreliminary/numofsamples.txt", stringsAsFactors = FALSE)
  colnames(numofsamples)<-c("ind", "samples")

  # Merge with invscores
  invscores<-merge(invscores, numofsamples)
  
  # Correction
  invscores$score_weighted<-invscores$score/invscores$samples

# Get individuals population
  popinfo<-read.table("data/use/avery_individuals/samples_population.txt", header = TRUE)
  
  # Get individual names
  popinfo$sample<-sub("^\\d+\\.","",popinfo$sample)
  popinfo$sample<-sub("nc", "NC", popinfo$sample, ignore.case = FALSE)
  
  # Merge
  invscores<-merge(invscores, popinfo[,c("sample", "super_pop")], by.x = "ind", by.y = "sample")
    
# Get individuals genotypes
  genotypes<-read.table("analysis/2020-05-23_07_imputationTables/new_genotypes.csv", header=TRUE)
  
  # Get individual names
  genotypes$Individual<-sub("^\\d+\\.","",genotypes$Individual)
  genotypes$Individual<-sub("nc", "NC", genotypes$Individual, ignore.case = FALSE)
  
  # Melt
  genotypes<-reshape2::melt(genotypes, id.vars="Individual" )
  colnames(genotypes)<-c("ind", "info", "genotype")
  
  # Merge
  invscores<-merge(invscores, genotypes)
  
# Cleanup
  rm(crossover_maps, crossover_inv, numofsamples, popinfo, genotypes)
```

 Fer gràfics de densitat per cada inversió ens sería útil per saber de forma descriptiva com es distribueixen els events de recombinació al voltant de les inversions. A més es separarien per genotips quan sigui possible, per tenir una primera visualització de l'efecte del genotip de la inversió sobre les taxes de recombinació locals.

```{r desnityPlots}

invscores$samplesize<-1
plotable<-aggregate(score_weighted ~ chr_w+start_w+end_w+genotype+info+start_i+end_i, data=invscores, FUN=mean)
plotable_v2<-aggregate(samplesize ~ chr_w+start_w+end_w+genotype+info+start_i+end_i, data=invscores, FUN=sum)

plotable<-merge(plotable, plotable_v2)

ggplot(plotable[plotable$genotype != ".",])+geom_rect(aes(xmin=start_w, xmax = end_w, ymin=log(0), ymax = log(score_weighted), fill = samplesize))+
  geom_vline(aes(xintercept=start_i), color = "red")+geom_vline(aes(xintercept=end_i), color = "red")+
  facet_grid(genotype~info, scales = "free")

```

Analitzar totes les inversions alhora és complicat perquè és un grup molt heterogeni, pero sí que es podria analitzar la relació entre els resultats obtinguts i diferents propietats de la inversió, per exemple, la correlació de la diferència entre taxes de recombinació de diferents genotips amb la mida de la inversió o amb si són NH/NAHR.


# Anlisi estadstic



Data was public. 
With bedtools intersect I took all crossovers that overlapped in any way with an inversion. I merged this information with the genotypes for each inversion and individual. 

```{r upload_files}

load("analysis/2020-05-06_09_crossovers/croxinv.Rdata")
crossovers_raw<-read.table("data/raw/avery_crossovers/allcrossovers_hg38.txt.gz", stringsAsFactors = FALSE, header = TRUE)
coords<-read.table("data/use/inversions_info/2020.02.inversions_hg38.gff", stringsAsFactors = FALSE, sep = "\t")
numofsamples<-read.table("report/2020-04-29_statisticalPreliminary/numofsamples.txt", stringsAsFactors = FALSE)
colnames(numofsamples)<-c("donor", "samples")


```

*Which is the resolution of crossovers, compared to inversion size?*
*Does the resolution change depending on chromosome?* --no
*Is the resolution of inversion-overlapping different?*
*Is the resolution in different genotypes different?*

```{r resolutions}

crossovers_raw$resolution<-crossovers_raw$pos.rightbound-crossovers_raw$pos.leftbound+1

summary(crossovers_raw$resolution)
ggplot(crossovers_raw)+geom_boxplot(aes(x = chr, y = log(resolution)))

coords$invsize<-coords$V6-coords$V3+1

summary(coords$invsize)
ggplot(coords)+geom_boxplot(aes(x = V2, y = log(invsize)))

summary(croxinv$resolution)
ggplot(croxinv)+geom_boxplot(aes(x = chromosome, y = log(resolution)))
```

First of all, I want a table with amount of recombination events per sperm sample overlapping with an inversion

```{r events_cell}

croxinv$individual<-as.character(croxinv$individual)
test<-data.frame(table(croxinv$individual, croxinv$inversio))


geno<-data.frame(table(croxinv$genotype, croxinv$inversion))

# Chromosome 21 test
chromosome<-"chr13"
test2<-crossovers_raw[crossovers_raw$chr == chromosome ,]
invs<-coords[coords$V1 == chromosome ,]
# invs
centel<-read.table("report/2020-04-29_statisticalPreliminary/repregions_hg38.txt")

invcoords<-croxinv[croxinv$chromosome == chromosome  & !duplicated(croxinv$inversion), ]
invcoords<-invcoords[!is.na(invcoords$inversion),]
invcoords$center<-invcoords$bp1s+(invcoords$bp2e-invcoords$bp1s)/2

library(ggforce)

a<-ggplot(test2)+geom_rect(aes(xmin=pos.leftbound, xmax=pos.rightbound, ymin = 0, ymax = 1), alpha = 0.002)+
  # geom_rect(data=invcoords, aes(xmin = bp1s, xmax = bp2e, ymin = 0.25, ymax = 0.75), fill = "red", color = "red", alpha = 0.5)+
  # geom_text(data=invcoords, aes(x = center, y = 0.8, label = inversion), angle = 90)+
  geom_rect(data=centel[centel$V1 == chromosome,], aes(xmin = V2, xmax = V3, ymin=0, ymax = 1), fill = "yellow", alpha = 0.3)+ggtitle(paste0(chromosome,", ",windowsize/1000, " kb window size"))+
  facet_zoom(xlim = c(8e7, 9e7) )

b<-ggplot(res)+geom_rect(aes(xmin=start, xmax=end,ymin=0,ymax=1, alpha=winsum))+theme(legend.position = "none")

png(paste0(chromosome,windowsize/1000,"kb.png"))
ggarrange(a,b, ncol=1)
dev.off()

test2$center<-test2$pos.leftbound+((test2$pos.rightbound-test2$pos.leftbound)/2)
test2$value<-1

windowsize<-100000
res<-data.frame(start=seq(8e7, 9e7, windowsize))
res$end<-res$start+windowsize
res$start<-res$start+1
res$winsum<-0


for (i in rownames(res)) {
  
  s<-res[i, "start"]
  e<-res[i, "end"]
  
  tmp<-test2[ (test2$pos.rightbound >= s & test2$pos.rightbound <=e)|
                (test2$pos.leftbound >= s & test2$pos.leftbound <=e)|
                (test2$pos.leftbound <= s & test2$pos.rightbound >=e) , ]

  tmp$fract.overlap<-0
  tmp$fract.overlap<-apply(tmp, 1, FUN = function(x){
        sum(seq(x["pos.leftbound"], x["pos.rightbound"],1) %in% seq(s,e,1))
      }
    )
  
  tmp$perc<-tmp$fract.overlap/tmp$resolution

  res[i,"winsum"]<-sum(tmp$perc)
  
}

tmp<-test2[ (test2$pos.rightbound >= s & test2$pos.rightbound <=e)|(test2$pos.leftbound >= s & test2$pos.leftbound <=e)| (test2$pos.leftbound <= s & test2$pos.rightbound >=e) , ]

tmp$fract.overlap<-0

x<-tmp[1,]

tmp$fract.overlap<-apply(tmp, 1, FUN = function(x){
        sum(seq(x["pos.leftbound"], x["pos.rightbound"],1) %in% seq(s,e,1))
      }
      )
tmp$perc<-tmp$fract.overlap/tmp$resolution
sum(tmp$perc)

res<-data.frame(donor=unique(tmp$donor))
res$start<-s
res$end<-e
res$window<-0

for( i in unique(res$donor)){
  res[ res$donor == i , "window"] <- sum(tmp[tmp$donor == i, "perc"])
}

res<-merge(res, numofsamples)

res$fraction<-res$window/res$samples


ggplot(test2)+geom_point(aes(x=pos.leftbound, y = cumsum(value), color = resolution ))


# Which is the chromosome with the most useful inversions? - chromosome 6 and chromosome 4
touse<-croxinv[!is.na(croxinv$genotype) & croxinv$genotype != "." & !is.na(croxinv$sample),]
table(touse$chromosome, touse$inversion)

# Now I make some summary of crossovers
crossovers_raw$buffer<-crossovers_raw$resolution
crossovers_raw$central.left<-crossovers_raw$pos.leftbound+crossovers_raw$buffer
crossovers_raw$central.right<-crossovers_raw$pos.rightbound-crossovers_raw$buffer

plot<-crossovers_raw[crossovers_raw$chr == chromosome,]
plot<-merge(plot, genotypes[genotypes$variable == "HsInv0266" & genotypes$value != ".",], by.x = "donor", by.y = "Individual")
table(data.frame(table(plot$value, plot$donor))$Var1,data.frame(table(plot$value, plot$donor))$Freq>1 )
chromosome<-"chr4"
plot$ponderation<-NA
plot[plot$value == "HET", "ponderation"]<-3/18
plot[plot$value == "INV", "ponderation"]<-2/18
plot[plot$value == "STD", "ponderation"]<-13/18

plot$ponderation_alpha<-plot$ponderation/10


invcoords<-touse[touse$chromosome == chromosome & !duplicated(touse$inversion), ]
library(ggforce)
ggplot(data=plot[plot$central.left > 185000000,])+geom_rect( aes(xmin=central.left, xmax=central.right, ymin = 0, ymax = 1, alpha = ponderation_alpha))+ 
  scale_alpha_continuous(range=c(min(plot$ponderation_alpha), max(plot$ponderation_alpha)))+
  geom_rect(data=invcoords[invcoords$inversion == "HsInv0266",], aes(xmin = bp1s, xmax = bp2e, ymin = 0.25, ymax = 0.75), color = "red", alpha = 0)+
  # geom_text(data=invcoords[invcoords$inversion == "HsInv0266",], aes(x = bp1s, y = 0.8, label = inversion))+
  # facet_zoom(xlim = c(30000000, 50000000) )+
  # facet_zoom(xlim = c(70000000, 100000000) )+
  # facet_zoom(xlim = c(150000000, 190000000) )
facet_grid(value ~.)+
  ggtitle("HsInv0266")




```



```{r test}

# 1 hacer ventanas 
multiplyer<-10
myinv<-"HsInv0095"
coords<-croxinv[croxinv$inversion == myinv & !is.na(croxinv$chromosome), c("chromosome", "bp1s", "bp2e")][1,]
buffer<-(coords$bp2e-coords$bp1s)
windows<-seq(from = coords$bp1s-(buffer*multiplyer), to = coords$bp2e+(buffer*multiplyer), length.out=(multiplyer*2)+2)

# 2 buscar los que solapan para cada ventana
here<-data.frame(windowstart=windows[1:length(windows)-1], windowend=windows[2:length(windows)] )
here$chromosome<-coords$chromosome
here$HETprob<-here$STDprob<-here$INVprob<-NA

for (row in rownames(here)) {
  myrow<-here[row,]
  tmp<-crossovers_raw[
                      ( (crossovers_raw$pos.leftbound >= myrow$windowstart 
                      & crossovers_raw$pos.rightbound <= myrow$windowend) |
                        (crossovers_raw$pos.leftbound <= myrow$windowstart 
                      & crossovers_raw$pos.rightbound >= myrow$windowend) |
                        (crossovers_raw$pos.leftbound >= myrow$windowstart 
                      & crossovers_raw$pos.leftbound <= myrow$windowend) |
                        (crossovers_raw$pos.rightbound >= myrow$windowstart 
                      & crossovers_raw$pos.rightbound <= myrow$windowend) )
                      & crossovers_raw$chr == myrow$chromosome,]
  counts<-data.frame(table(tmp$donor))
  colnames(counts)<-c("donor", "crossovers")
  counts<-merge(counts, numofsamples, all=TRUE)
  counts[is.na(counts$crossovers),"crossovers"]<-0
  counts<-merge(counts, genotypes[genotypes$variable == myinv,], by.x = "donor", by.y = "Individual")
  counts<-counts[counts$value != ".",]
  het<-sum(counts[counts$value == "HET","crossovers"])/sum(counts[counts$value == "HET","samples"])
  std<-sum(counts[counts$value == "STD","crossovers"])/sum(counts[counts$value == "STD","samples"])
  inv<-sum(counts[counts$value == "INV","crossovers"])/sum(counts[counts$value == "INV","samples"])
  here[row, c("HETprob", "STDprob", "INVprob")]<-c(het, std, inv)
}

plot<-melt(here, id.vars = c("windowstart", "windowend", "chromosome"))
plot$inversion<-NA
plot[plot$windowstart == coords$bp1s, "inversion"]<-myinv

ggplot(plot)+geom_bar(aes(x=windowstart, y = value, fill = inversion), stat = "identity")+
  facet_grid(.~variable)+ggtitle("myinv")+
  theme(legend.position = "none")

# anova
library(tidyverse)
library(ggpubr)
library(rstatix)

  myinv<-"HsInv0045"
  coords<-croxinv[croxinv$inversion == myinv & !is.na(croxinv$chromosome), c("chromosome", "bp1s", "bp2e")][1,]

  tmp<-croxinv[croxinv$inversion == myinv,]

  counts<-data.frame(table(tmp$individual))
  colnames(counts)<-c("donor", "crossovers")
  counts<-merge(counts, numofsamples, all=TRUE)
  counts[is.na(counts$crossovers),"crossovers"]<-0
  counts<-merge(counts, tmp[!duplicated(tmp$individual), c("individual","genotype")], by.x = "donor", by.y = "individual")
  counts<-counts[counts$genotype != ".",]
  
  counts$weight<-counts$crossovers/counts$samples
  
  counts$genotype<-factor(counts$genotype, levels = c("STD", "INV", "HET"))

  ggboxplot(counts, x = "genotype", y = "weight", title = myinv)

  # Outliers
  
  counts %>% 
  group_by(genotype) %>%
  identify_outliers(weight)

  # Normality assumptions
  # Build the linear model
model  <- lm(weight ~ genotype, data = counts)
# Create a QQ plot of residuals
ggqqplot(residuals(model))

# global shapiro test
shapiro_test(residuals(model))

# groups shapiro test
counts %>%
  group_by(genotype) %>%
  shapiro_test(weight) # no sale..

# groups qqplot
ggqqplot(counts, "weight", facet.by = "genotype")


# homogeneity of variance
plot(model, 1)

counts %>% levene_test(weight ~ genotype)

# anova

res.aov <- counts %>% anova_test(weight ~ genotype)
res.aov



```


# Gene conversion test

```{r}

test<-read.table("data/raw/avery_crossovers/allcrossovers_hg38.txt.gz", sep ="\t", header = TRUE)

test$size<-test$pos.rightbound-test$pos.leftbound

summary(test$size<500)
```
