<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01//EN"
"http://www.w3.org/TR/html4/strict.dtd">
<html>
	<head>

		<script type="text/javascript" src="js/jquery-1.4.2.min.js"></script>
		<script type="text/javascript" src="js/jquery.ba-hashchange.min.js"></script>
		<!-- Google Analytics tracking code-->
		<script type="text/javascript">
			var _gaq = _gaq || [];
			_gaq.push(['_setAccount', 'UA-16521993-1']);
			_gaq.push(['_trackPageview']);

			(function() {
				var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
				ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
				var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
			})();
		</script>
		<link href="style/css/default.css" rel="stylesheet" type="text/css">
		<title>QCTOOL v2</title>
	</head>
	<body>
		<div class="header" id="header1">
			<!--<img src = "style/images/qctool.png" class = "header_img" ></img>-->
			<div class="header_text">
				qctool v2
			</div>
			<div class="header_subtext">
				A tool for quality control and analysis of gwas datasets.
			</div>
		</div>
		<div class="button_bar">
			<div class="nav_button" name="overview">
				<a href="index.html" class="nav_button_content">overview</a>
			</div>
			<div class="nav_button" name="documentation">
				<a href="reference.html" class="nav_button_content">documentation</a>
			</div>
			<div class="nav_button" name="tutorial">
				<a href="examples.html" class="nav_button_content">examples</a>
			</div>
			<div class="nav_button" name="file_formats">
				<a href="file_formats.html" class="nav_button_content">file formats</a>
			</div>
			<div class="nav_button" name="download">
				<a href="download.html" class="nav_button_content">download</a>
			</div>
		</div>
		<div class="main_text">
			<div id="documentation">
				<div class="display_diagram_right">
					<object data="style/images/qctool_cheatsheet.svg" width="328" height="800" type="image/svg+xml" align="center">
						</object>
				</div>
				<h3>
					Introduction
				</h3>
				<p>
					Data flows through QCTOOL in a fixed pipeline that is shown in the schematic on the right. In brief, data for one or more cohorts is read into QCTOOL one variant at a time, merged into one combined dataset, and undergoes further processing. This combined dataset is then passed to output modules and various analyses. Command-line options (small grey text in the diagram on the right) affect how QCTOOL reads, filters, processes, and outputs data.
				</p>
				<p>
					A full description for all options is always available via the command
				</p>
				<pre>
$ qctool -help
</pre>
				<p>
					which produces <a href="qctool_help.txt">this output</a>.
				</p>
				<p>
					Some basic functions that QCTOOL can be used for are: to convert files between formats, and to compute per-SNP or per-sample summary statistics. The rest of this page documents these functions, for further examples see the <a href='examples.html'>examples page</a>.
				</p>
				<h3>
					Converting between file formats
				</h3>
				<p>
					A basic command to convert between file formats is:
				</p>
				<div>
					<code>qctool -g &lt;input file&gt; -og &lt;output file&gt;</code>
				</div>
				<p>
					This command reads genotypes and/or other data from the input file(s) and outputs them to the output file. A considerable number of file formats are now supported by this tool - see the <a href='file_formats.html'>file formats page</a> for a full list. For the most common formats, the filename extension is used to determine the file format. Where this doesn't happen automatically the <code>-filetype</code> (for input files) and <code>-ofiletype</code> (for output files) options can be used to specify the filetype.
				</p>
				<p>
					The option <code>-og -</code> means "output to standard output".
				</p>
				<p>
					Note that although it's not required to supply a sample file (<code>-s</code> option), for some combinations of input/output file formats (such as GEN to VCF) supplying a sample file is preferable to ensure that the output files have appropriate sample labels.
				</p>
				<div>
					<code>qctool \<br>
					-g &lt;input file&gt; \<br>
					-s &lt;sample file 1&gt; \<br>
					-og &lt;output file&gt;</code>
				</div>
				<p>
					As shown on the right, a number of options affect the format of the genotype output file.
				</p>
				<ul>
					<li>
						<code>-ofiletype</code>: this determines the output file format; if not suplied a guess is made based on the filename as described <a href='file_formats.html'>here</a>.
					</li>
					<li>
						<code>-omit-chromosome</code>: (for GEN and SHAPEIT files) this tells QCTOOL to write files without an initial chromosome column.
					</li>
					<li>
						<code>-vcf-output-fields</code>: (for VCF file output) this tells QCTOOL to restrict the set of FORMAT fields written to the output file to the specified list.
					</li>
					<li>
						<code>-bgen-bits</code>: (for BGEN file output) this tells QCTOOL to use the specified number of bits when encoding probabilities in BGENv1.2 format.
					</li>
				</ul>
				<h3>
					Sorting data
				</h3>
				<p>
					The <code>-sort</code> option tells QCTOOL to sort the variants in the output file. QCTOOL does this by first writing the output file unsorted, but internally keeping a record of the file locations of the data for each variant. Then the file is renamed to a temporary file and copied back to the destination in the sorted order.
				</p>
				<p>
					The sort command uses the <code>-compare-variants-by</code> option to determine what to sort by. This defaults to '<code>position,rsid,SNPID,alleles</code>' which means variants are sorted by genomic position, ID fields, and alleles. (The <code>-compare-variants-by</code> option is used by several parts of QCTOOL - whenever variants need to be compared or ordered.)
				</p>
				<p>
					Sorting is only supported for those output file types that support efficient seeking - in particular sorting will not work with gzipped output formats. (The BGEN format does support sorting.)
				</p>
				<h3>
					Merging data
				</h3>
				<p>
					QCTOOL accepts multiple occurences of the <code>-g</code> option specifying multiple cohorts worth of data. QCTOOL merges these datasets internally, forming a combined dataset with <em>&amp;Sum;<sub>i</sub> N<sub>i</sub></em> samples, where <em>N<sub>i</sub></em> is the number of samples in cohort <em>i</em>.
				</p>
				<p>
					<b>Important.</b> QCTOOL does not do an exhaustive search to try to match variants between cohorts (in fact it traverses each input file exactly once from start to finish). Instead, it uses the following algorithm to merge variants:
				</p>
				<ol>
					<li>For each variant in cohort 1;
					</li>
					<li>Look for the next matching variant in cohorts 2, ..., n; but stop searching when a variant with greater position is found;
					</li>
					<li>If a match is found, merge the data and process it;
					</li>
					<li>Otherwise, omit this variant and move to the next.
					</li>
				</ol>
				<p>
					Variants are matched using the fields specified by the <code>-compare-variants-by</code> option (and in the order they're specified). (The genomic position must always be the first field sorted on.)
				</p>
				<p>
					The implication of the above is that the merge will work best when either
				</p>
				<ul>
					<li>Data in each cohort is sorted (by the same fields specified in <code>-compare-variants-by</code>); or
					</li>
					<li>Data in each cohort has exactly the same set of SNPs in exactly the same order.
					</li>
				</ul>
				<p>
					In other situations the result of QCTOOL may not be a full join, i.e. it may omit variants that could in principle be merged.
				</p>
				<h3>
					Computing summary statistics
				</h3>
				<p>
					QCTOOL will compute per-variant and per-sample summary statistics of several different types. In this version of QCTOOL, all per-variant quantities are output to the output file specified by <code>-osnp</code>, and all per-sample quantities are output to the output file specified by <code>-osample</code>. These options will choose a file type based on format of their argument according to the following table:
				</p>
				<table>
					<tr>
						<th>
							Option
						</th>
						<th>
							File format
						</th>
					</tr>
					<tr>
						<td>
							<code>&lt;filename&gt;.csv</code>
						</td>
						<td>
							CSV file.
						</td>
					</tr>
					<tr>
						<td>
							<code>&lt;filename&gt;.tsv</code>
						</td>
						<td>
							Tab-separated file.
						</td>
					</tr>
					<tr>
						<td>
							<code>sqlite://&lt;filename&gt;:&lt;table name&gt;</code>
						</td>
						<td>
							A specified table in a <a href="http://www.sqlite.org">sqlite3</a> file.
						</td>
					</tr>
					<tr>
						<td>
							<code>&lt;filename&gt;.txt</code> (or anything else)
						</td>
						<td>
							Space-separated file
						</td>
					</tr>
					<tr>
						<td>
							<code>-</code>
						</td>
						<td>
							Write to standard output
						</td>
					</tr>
				</table>
				<p>
					A number of options produce data that is output to the file specified by <code>-osnp</code>. They are tabulated in the following table.
				</p>
				<table>
					<tr>
						<th>
							Option
						</th>
						<th>
							Summary
						</th>
					</tr>
					<tr>
						<td>
							<code>-snp-stats</code>
						</td>
						<td>
							Basic per-variant summary statistics - frequency, missing proportion, P-value for Hardy-Weinberg, etc.
						</td>
					</tr>
					<tr>
						<td>
							<code>-intensity-stats</code>
						</td>
						<td>
							Mean per-variant X and Y channel intensities.
						</td>
					</tr>
					<tr>
						<td>
							<code>-fit-clusters</code>
						</td>
						<td>
							Fit multivariate T distribution clusters to genotype calls and intensities, output parameters.
						</td>
					</tr>
					<tr>
						<td>
							<code>-annotate-sequence</code>
						</td>
						<td>
							Annotation of reference base(s) and flanking sequence from a reference FASTA file.
						</td>
					</tr>
					<tr>
						<td>
							<code>-annotate-genetic-map</code>
						</td>
						<td>
							Annotation of interpolated genetic map positions from genetic map file(s).
						</td>
					</tr>
					<tr>
						<td>
							<code>-annotate-bed3</code>,<code>-annotate-bed4</code>
						</td>
						<td>
							Annotation of occupancy of regions defined in one or more BED files.
						</td>
					</tr>
				</table>
				<h3>
					Details of summary statistics
				</h3>
				<p>
					The following table lists columns output by QCTOOL when using each option. Note that QCTOOL was originally designed for human data; summary statistic computation currently only supports diploid samples (and haploids on the sex chromosomes, as determined by sex information in the sample file). Summary statistics currently only support biallelic variants.
				</p>
				<p>
					Computed by <code>-snp-stats</code>:
				</p>
				<dl class="column_definition">
					<dt>
						<code>[A|B|AA|AB|BB|NULL]</code>
					</dt>
					<dd>
						Counts of each haploid or diploid genotype. For hard-called genotype data, or when using <code>-threshold</code>, these will be integers. Otherwise they reflect expected counts and may have nonzero fractional part.
					</dd>
					<dt>
						<code>incorrect_ploidy</code>
					</dt>
					<dd>
						<em>(Only computed on sex chromosomes)</em>.The count of incorrectly coded genotypes for haploid samples. Currently, ploidy is assumed to be diploid on the autosomal chromosomes, diploid/haploid for females/males on the X chromosome, and 0-ploid / haploid for females and males on the Y chromosome. The coding of haploid genotypes is detemined by the <code>-haploid-genotype-coding</code> option. The default is to assume haploid samples are coded as homozygous genotypes.
					</dd>
					<dt>
						<code>unknown_ploidy</code>
					</dt>
					<dd>
						<em>(Only computed on sex chromosomes)</em>: The count of samples whose ploidy could not be determined. On the X and Y chromosomes, this occurs for any sample with missing sex information (as determined by the column of the sample file specified by <code>-sex-column</code>).
					</dd>
					<dt>
						<code>alleleA_frequency</code>, <code>alleleB_frequency</code>, <code>minor_allele_frequency</code>
					</dt>
					<dd>
						The frequency of the A and B alleles, and of the minor (less frequent) allele
					</dd>
					<dt>
						<code>missing_call_proportion</code>
					</dt>
					<dd>
						The proportion of missing calls (defined as genotypes with highest genotype probability less than 0.9).
					</dd>
					<dt>
						<code>missing_proportion</code>
					</dt>
					<dd>
						The proportion of missing data (defined as the sum of NULL genotype probabilities, divided by the total number of samples). An imputed variant will usually have a value of close to zero here. A directly-typed SNP may have nonzero NULL call probability e.g. if called using a clustering algorithm.
					</dd>
					<dt>
						<code>impute_info</code>
					</dt>
					<dd>
						The IMPUTE info metric. This is defined as one minus the expected variance of a genotype, divided by the variance of a genotype with the estimated allele frequency under Hardy-Weinberg. This is the traditional metric which is always computed as though for diploid samples and does not take into account missing data. It has the property that <em>info &amp;leq; 1</em> with equality iff all the genotype calls are completely certain.
					</dd>
					<dt>
						<code>info</code>
					</dt>
					<dd>
						The (modified) info metric, taking account non-diploid samples (e.g. males on the X chromosome) and missing data. This info measure has the property that <em>info &amp;leq; 1 - missingness</em> with equality iff all the genotype calls are completely certain.
					</dd>
					<dt>
						<code>HW_exact_p_value</code>, <code>HW_lrt_p_value</code>
					</dt>
					<dd>
						P-values for Hardy-Weinberg equilibrium. The first is computed using the <a href="http://www.sph.umich.edu/csg/abecasis/Exact/index.html">exact test</a> of Wigginton et al, "<em>A Note on Exact Tests of Hardy-Weinberg Equilibrium</em>", Wigginton et al, <i>Am. J. Hum. Genet</i> (2005) 76:887-93. The second is computed using an asymptotically chi-squared distributed likelihood ratio statistic. (Note the latter may become invalid when counts are small.)
					</dd>
					<dt>
						<code>HW_females_exact_pvalue</code>, <code>HW_females_lrt_pvalue</code>
					</dt>
					<dd>
						<em>(Only computed on X chromosome)</em>: P-values for Hardy-Weinberg equilibrium in females.
					</dd>
					<dt>
						<code>male_female_exact_pvalue</code>
					</dt>
					<dd>
						<em>(Only computed on the X chromosome)</em> A P-value for equal allele frequency in males and females.
					</dd>
					<dt>
						<code>male_female_and_HW_lrt_pvalue</code>
					</dt>
					<dd>
						<em>(Only computed on the X chromosome)</em> A P-value for Hardy-Weinberg in females and equal allele frequency in males and females, computed using asymptotically chi-squared distributed likelihood ratio statistic. (Note this P-value may be invalid when counts are small.)
					</dd>
				</dl>
				<p>
					Computed by <code>-intensity-stats</code>:
				</p>
				<dl class="column_definition">
					<dt>
						<code>mean_X</code>, <code>mean_Y</code>
					</dt>
					<dd>
						Mean X channel and Y channel intensities
					</dd>
					<dt>
						<code>var_X</code>, <code>var_Y</code> <code>cov_XY</code>
					</dt>
					<dd>
						Variance and covariance in X and Y channel intensities
					</dd>
					<dt>
						<code>g=&lt;x&gt;:mean_X</code>, <code>g=&lt;x&gt;:mean_Y</code>
					</dt>
					<dd>
						Mean X channel and Y channel intensities for calls having the given genotype
					</dd>
					<dt>
						<code>g=&lt;x&gt;:var_X</code>, <code>g=&lt;x&gt;:var_Y</code>, <code>g=&lt;x&gt;:cov_XY</code>
					</dt>
					<dd>
						Variance and covariance in X and Y channel intensities for calls having the given genotype
					</dd>
				</dl>
				<p>
					Computed by <code>-compare-to &lt;auxiliary genotype file&gt; &lt;auxiliary sample file&gt;</code>:
				</p>
				<dl class="column_definition">
					<dt>
						<code>compared_variant_rsid</code>, <code>compared_variant_alleleA</code>, <code>compared_variant_alleleB</code>
					</dt>
					<dd>
						RSID and first and second alleles of the variant in the auxiliary data this variant was compared to. (Matching of variants is controlled by the <code>-compare-variants-by</code> option.)
					</dd>
					<dt>
						<code>&lt;sample ID&gt;(&lt;ID in primary data&gt;~&lt;ID in compared data&gt;):genotype</code>
					</dt>
					<dd>
						Dosage for the given sample in the primary data and in the compared data. (Matching of samples is controlled by the <code>-match-samples-by</code> column.)
					</dd>
					<dt>
						<code>pairwise_non_missing_calls</code>
					</dt>
					<dd>
						The number of genotypes that were non-missing in primary and compared datasets
					</dd>
					<dt>
						<code>pairwise_concordant_calls</code>, <code>concordance</code>
					</dt>
					<dd>
						The number and proportion of pairwise non-missing genotypes that were concordant in primary and compared datasets
					</dd>
					<dt>
						<code>concordance</code>
					</dt>
					<dd>
						The number of pairwise non-missing genotypes that were concordant in primary and compared datasets
					</dd>
					<dt>
						<code>correlation</code>
					</dt>
					<dd>
						The correlation between genotype dosages in the primary and compared datasets
					</dd>
				</dl>
				<p>
					Computed by <code>-annotate-bed[3|4] &lt;name of file in <a href="https://genome.ucsc.edu/FAQ/FAQformat.html#format1">BED</a> format&gt;</code>:
				</p>
				<dl>
					<dt>
						<code>&lt;filename&gt;</code> (minus the extension)
					</dt>
					<dd>
						For <code>-annotate-bed3</code>, a 0/1 indicator of whether the variant lies in an interval in the given BED file; for <code>-annotate-bed4</code>, a list of values from the fourth column of the BED file for intervals which the variant lies in. Note positions in BED files are in 0-based coordinates; QCTOOL assumes 1 based coordinates for the positions of variants and will translate BED coordinates accordingly.
					</dd>
				</dl>
				<p>
					Computed by <code>-annotate-sequence &lt;name of fasta file&gt; &lt;name&gt;</code>:
				</p>
				<dl>
					<dt>
						<code>&lt;name&gt;_alleleA</code>
					</dt>
					<dd>
						The content of the given FASTA file for the bases covered by the first allele at each variant
					</dd>
					<dt>
						<code>&lt;name&gt;_alleleA_left_flanking</code>, <code>&lt;name&gt;_alleleA_right_flanking</code>
					</dt>
					<dd>
						The left and right flanking sequence from the given FASTA file. The length of flanking sequence output is controlled by the <code>-flanking</code> option.
					</dd>
					<dt>
						<code>&lt;name&gt;_alleleB_if_indel</code>
					</dt>
					<dd>
						The content of the given FASTA file for the bases covered by the second allele. This is only output if the length of the two alleles is different (i.e. variant is an insertion or deletion).
					</dd>
				</dl>
			</div>
		</div>
	</body>
</html>
