---
title: "R Notebook"
output: html_notebook
editor_options: 
  chunk_output_type: console
---

```{r, eval=FALSE}

# To install required packages (did not work)

if (!requireNamespace("BiocManager", quietly=TRUE)){ install.packages("BiocManager") }

BiocManager::install("GenomicRanges")
BiocManager::install("ggbio")
library(GenomicRanges)
library(devtools)

install_git("git://github.com/daewoooo/primatR.git", branch = "master")
library(git2r)

```

```{r hotspotter}

#' Find hotspots of genomic events
#'
#' Find hotspots of genomic events by using kernel \link{density} estimation.
#'
#' The hotspotter uses \code{\link[stats]{density}} to perform a KDE. A p-value is calculated by comparing the density profile of the genomic events with the density profile of a randomly subsampled set of genomic events. Due to this random sampling, the result can vary for each function call, most likely for hotspots whose p-value is close to the specified \code{pval}.
#' 
#' @param gr A \code{\link{GRanges-class}} object containing the coordinates of the genomic events.
#' @param bw Bandwidth used for kernel density estimation (see \code{\link[stats]{density}}).
#' @param pval P-value cutoff for hotspots.
#' @param num.trial A number of randomly subsampled set of genomic events.
#' @return A \code{\link{GRanges-class}} object containing coordinates of hotspots with p-values.
#' @importFrom stats density runif ecdf
#' @importFrom S4Vectors endoapply
#' @import GenomicRanges
#' @author Aaron Taudt
#' @export

hotspotter <- function(gr, bw, pval=1e-8, num.trial=100) {

  set.seed(123) # fix seed for random permutations of bootstrapping
  
  ## Iterate over chromosomes and calculate p-values
  pranges.list <- GenomicRanges::GRangesList()
  for (chrom in seqlevels(gr)) {
    grc <- gr[seqnames(gr)==chrom]
    if (length(grc)>1) {
      midpoints <- (start(grc)+end(grc))/2
      # ~~  ADDITION
      if (bw == "nrd"){
        bw<-bw.nrd(midpoints)
      }
      # ~~  END OF ADDITION
      kde <- stats::density(midpoints,bw=bw,kernel='gaussian')
      # Random distribution of genomic events
      kde.densities <- numeric()
      for (i1 in seq_len(num.trial)) {
        midpoints.r <- round(stats::runif(length(midpoints),1,seqlengths(gr)[chrom]))
        kde.r <- stats::density(midpoints.r,bw=bw,kernel='gaussian')
        kde.densities <- c(kde.densities, kde.r$y)
      }
      # Use ecdf to calculate p-values 
      p <- 1-stats::ecdf(kde.densities)(kde$y)
      pvalues <- data.frame(chromosome=chrom,start=kde$x,pvalue=p)
      # Make GRanges
      pvalues$end <- pvalues$start
      pvalues$chromosome <- factor(pvalues$chromosome, levels=seqlevels(gr))
      pvalues <- as(pvalues,'GRanges')
      seqlevels(pvalues) <- seqlevels(gr)
      suppressWarnings(
        seqlengths(pvalues) <- seqlengths(gr)[names(seqlengths(pvalues))]
      )
      # Resize from pointsize to bandwidth
      suppressWarnings(
        pvalues <- GenomicRanges::resize(pvalues, width=bw, fix='center')
      )
      pvalues <- trim(pvalues)
      ## Find regions where p-value is below specification
      mask <- pvalues$pvalue <= pval
      rle.pvals <- rle(mask)
      rle.pvals$values <- cumsum(rle.pvals$values+1)
      pvalues$group <- inverse.rle(rle.pvals)
      if (length(which(mask))>0) {
        pvalues.split <- split(pvalues[mask],pvalues$group[mask])
        pranges <- unlist(endoapply(pvalues.split, function(x) { y <- x[1]; end(y) <- end(x)[length(x)]; y$pvalue <- min(x$pvalue); return(y) }))
        pranges$group <- NULL
        pranges$num.events <- GenomicRanges::countOverlaps(pranges, grc)
        ## Make sure only non-zero counts are reported
        pranges <- pranges[pranges$num.events > 0]
        pranges.list[[chrom]] <- pranges
      }
    }
  }
  pranges <- unlist(pranges.list, use.names=FALSE)
  names(pranges) <- NULL
  
  return(pranges)
}

```

```{r runHotspotter}
gr <- GRanges(
    seqnames = Rle(c("chr1", "chr2", "chr1", "chr3"), c(1, 3, 2, 4)),
    ranges = IRanges(101:110, end = 111:120, names = head(letters, 10)),
    strand = Rle(strand(c("-", "+", "*", "+", "-")), c(1, 2, 2, 3, 2)),
    score = 1:10,
    GC = seq(1, 0, length=10))
gr



df
gr<-makeGRangesFromDataFrame(df)  # strand value "." is replaced with "*"
granges(gr)
mcols(gr)
seqnames(gr)

seqlengths(gr)<-chroms$V5[1:3]
# ----------------
cO<-read.table("../../data/use/avery_crossovers/allcrossovers_center.bed", header = T)
chroms<-read.table("../../data/use/assembly_hg38/cromlength.gff")
chromlengths<-chroms$V5
names(chromlengths)<-chroms$V1
cO.GR<-makeGRangesFromDataFrame(cO, ignore.strand = T, start.field = "pos.leftbound", end.field = "pos.rightbound",seqinfo = chromlengths )


# cO.GR@seqnames

mcols(cO.GR)

# seqlengths(cO.GR)<-chroms$V5



hotspots<-hotspotter(cO.GR, bw=2000, num.trial=1000)

library(ggbio)


autoplot(hotspots, layout = "karyogram")

invcoords<-read.table("../../data/use/inversions_info/2020.07.inversions_hg38.csv", sep = "\t")

recInf<-read.table("../../data/use/inversions_info/2020-08_inversionInfo_hg38.csv", sep = "\t", header  =TRUE)


invcoords<-merge(invcoords, recInf, by.x = "V2", by.y = "Inversion" )
invcoords$chromosome<-paste0("chr",invcoords$V1)

invcoords<-invcoords[!(invcoords$chromosome %in% c("chrX", "chrY")), c("V2", "chromosome", "V3", "V6", "Origin")]

invs.gr<-makeGRangesFromDataFrame(invcoords, ignore.strand = T, start.field = "V3", end.field = "V6", seqnames.field = "chromosome", seqinfo = chromlengths, keep.extra.columns = T)

autoplot(invs.gr, layout = "karyogram", aes(color = Origin, fill = Origin))

# hotspots2k1k<-
  
  autoplot(hotspots, layout = "karyogram")+
layout_karyogram(data = invs.gr, geom = "rect", ylim = c(0, 10), aes(color = Origin, fill = Origin)) 

# hotspots200100<-autoplot(hotspots, layout = "karyogram")+
# layout_karyogram(data = invs.gr, geom = "rect", ylim = c(0, 10), aes(color = Origin, fill = Origin)) 


autoplot(cO.GR, layout = "karyogram", alpha = 0.007)+
layout_karyogram(data = invs.gr, geom = "rect", ylim = c(0, 10), aes(color = Origin, fill = Origin)) 




```


```{r artificialRecombinationEvents}




```
