---
title: "Inversion filtering analysis"
author: "Ruth GÃ³mez Graciani"
---

```{r, setup, include=FALSE}
# require("knitr")
# require("tidyr")
# require("ggforce")
# require("ggpubr")
# require("reshape2")
# library("ggplot2")
# require("stringr")
knitr::opts_knit$set(root.dir = "~/20200401_RealRecombination/")
# outdir<-"report/2020-08-20_defMethods/"

say.goods<-function(vector){
  sum(vector == "PASS")
}


```

We started with the following list of inversions. When (n = m) inversions is stated, it  means there are n inversions in the list, of which m are unique inversions and n-m are repeated because they had DEL genotypes. 

INVERIONS STEP01: autosomal inversions (106 = 104)
INVERIONS STEP02: inversions with imputability info (111 = 109)
INVERIONS IMPUTE2 = STEP01 int STEP02: autosomal invs with imputability info (94 = 92)
INVERIONS STEP03: is imputable or has tag SNPs or is non polymorphic in at least one population (93 = 92)
INVERIONS STEP04 = STEP03 int. STEP01: imputable autosomal inversions (83 = 82)


```{r}

# LOAD DATA

input.invs<-read.table("data/use/inversions_info/Inversions_imputability.txt", header = TRUE)
input.invs$STEP02.has.imp.info<-"PASS"
coord.invs<-read.table("data/use/inversions_info/Inversions_coordinates_hg19.csv", col.names = c("Inversion", "chromosome", "B1s19", "B1e19", "B2s19", "B2e19"))
coord.invs<-coord.invs[coord.invs$Inversion != "HsInv0052",]

# STEP01: autosomal

coord.invs$STEP01.is.autosomal<-"FILTERED"
coord.invs[!(coord.invs$chromosome %in% c("chrX", "chrY")),"STEP01.is.autosomal" ]<-"PASS"

say.goods(coord.invs$STEP01.is.autosomal)

# STEP02: imputability info

test<-merge(coord.invs[, c("Inversion", "chromosome","STEP01.is.autosomal")], input.invs , all = TRUE)
test[is.na(test$STEP02.has.imp.info), "STEP02.has.imp.info"]<-"FILTERED"
test[test$Inversion == "HsInv0822", c("chromosome", "STEP01.is.autosomal")]<-c("chrX", "FILTERED")

say.goods(test$STEP02.has.imp.info)

# IMPUTE2
test$IMPUTE2<-"FILTERED"
test[test$STEP01.is.autosomal == "PASS" & test$STEP02.has.imp.info == "PASS","IMPUTE2"]<-"PASS"
say.goods(test$IMPUTE2)

# STEP03: imputable
test$STEP03.tag.or.imputable<-"FILTERED"
accepted_tags<-c("Tagged", "Imputable", "No_Polymorphic")
test[test$GLB %in% accepted_tags | test$EUR %in% accepted_tags | test$AFR %in% accepted_tags , "STEP03.tag.or.imputable"]<-"PASS"

say.goods(test$STEP03.tag.or.imputable)

# STEP04: imputable autosomal

test$STEP04.autosomal.imputable<-"FILTERED"
test[(test$STEP03.tag.or.imputable == "PASS" & test$STEP01.is.autosomal == "PASS" ), "STEP04.autosomal.imputable"] <- "PASS"

say.goods(test$STEP04.autosomal.imputable)


```


After the imputation, the genotypes were filtered differently depending on the individual's population. The amount of inversions analyzed with IMPUTE2 is 94 = 92 (autosomal inversions with imputability info), so the table has 1880 rows (20 individuals x 94 inversions).

```{r createGenotypesGOOD,  echo = FALSE, eval=FALSE}

 # Individual genotypes
   # Get individuals genotypes
    genotypes<-read.table("analysis/2020-05-23_07_imputationTables/imputationResults.csv", header=TRUE, sep = ",")
  
    # Get individual names
    genotypes$Individual<-sub("^\\d+\\.","",genotypes$Individual)
    genotypes$Individual<-sub("nc", "NC", genotypes$Individual, ignore.case = FALSE)
    genotypes$Individual<-sub("ab", "", genotypes$Individual, ignore.case = FALSE)


```

Genotypes filtering steps:

GENOTYPES STEP01: Inversions marked as imputable and tagged were accepted. SAS, ALL and EAS individuals were filtered according to the GLB imputability column. EUR and AFR were filtered according to their corresponging columns. Inversions HsInv0191 and HsInv1075 were not accepted for SAS and ALL individuals because it was observed in AFR and EUR individuals that good quality (all results with probability >0.8) not always implied good uniformity (all 'global' results equal and equal to 'same population' control). 

GENOTYPES STEP02: Quality filter. Individuals without a same population control result (mostly ALL and SAS, but some EAS as well), were required to have probabilities > 0.8 in all the global results. Otherwise, the same population control probability was required to be >0.8.

GENOTYPES STEP03:  Uniformity filter. All genotypes for EUR and AFR individuals were accepted regardless their uniformity, because we trust the same population control provided that the corresponging probability is >0.8. Other populations' genotypes were accepted when all their available measurements predicted the same genotype (all 4 measurements in most EAS individuals and the global results only in most SAS and ALL individuals).

GENOTYPES STEP04: for a genotype to be accepted, it must pass all three filters. The accepted genotype is the same population control when available, and the global results when the same population control is missing.


```{r}

# Now I apply one by one the filters to make sure they are correct
# Things I noted that could add some inversions: take into account non polymorphic ones

# STEP01: imputability 
genotypes$STEP01.pop.imputability<-"FILTERED"
accepted_tags<-c("Tagged", "Imputable")

genotypes[genotypes$Population %in% c("SAS", "ALL", "EAS") & genotypes$GLB %in% accepted_tags, "STEP01.pop.imputability"  ]<-"PASS"
genotypes[genotypes$Population %in% c("AFR") & genotypes$AFR %in% accepted_tags, "STEP01.pop.imputability"  ]<-"PASS"
genotypes[genotypes$Population %in% c("EUR") & genotypes$EUR %in% accepted_tags, "STEP01.pop.imputability"  ]<-"PASS"
genotypes[genotypes$Inversion %in% c("HsInv0191", "HsInv1075") & genotypes$Population %in% c("SAS", "ALL"),"STEP01.pop.imputability"]<-"FILTERED"

# STEP02: quality
genotypes$STEP02.quality<-"FILTERED"

genotypes[is.na(genotypes$Probability_con)&(genotypes$Probability_100 > 0.8 & genotypes$Probability_250 > 0.8 & genotypes$Probability_500 > 0.8) , "STEP02.quality"]<-"PASS"

genotypes[!is.na(genotypes$Probability_con)&genotypes$Probability_con > 0.8, "STEP02.quality" ]<-"PASS"

# STEP03: uniformity
genotypes$STEP03.uniformity<-"FILTERED"
genotypes[genotypes$Population %in% c("EUR", "AFR") ,"STEP03.uniformity"] <- "PASS"
genotypes[genotypes$Uniformity == "good", "STEP03.uniformity"] <-"PASS"

genotypes[genotypes$Genotype_500 == "?","Genotype_500" ]<-NA
genotypes$Genotype_500<-factor(genotypes$Genotype_500)

genotypes[is.na(genotypes$Genotype_con) & (genotypes$Genotype_100 == genotypes$Genotype_250 & genotypes$Genotype_250 == genotypes$Genotype_500), "STEP03.uniformity"] <-"PASS"


# STEP04: all together
genotypes$STEP04.general.filter<-"FILTERED"
genotypes[genotypes$STEP01.pop.imputability == "PASS" &
            genotypes$STEP02.quality == "PASS" &
            genotypes$STEP03.uniformity == "PASS" ,"STEP04.general.filter" ]<-"PASS"

# Accepted genotype

genotypes$accepted_genotype<-genotypes$Genotype_con
genotypes[is.na(genotypes$accepted_genotype), "accepted_genotype"]<-genotypes[is.na(genotypes$accepted_genotype), "Genotype_500"]

```

Once the genotypes are filtered, they are aggregated and counted to filter inversions by the amount of different haplotypes available. 

INVERIONS STEP05: those inversions with more than 3 individuals genotyped and with both heterozygous and homozygous individuals available (77 = 76).

INVERIONS STEP06 = STEP04 int. STEP05: autosomal imputable good quality genotyped inverisons (68 = 67)

```{r}

   # STEP05: AGGREGATING  GENOTYPES
    
    counts<-data.frame(unclass(table(as.character(genotypes$Inversion), genotypes$accepted_genotype)))
    counts$HOMO <- counts$INV + counts$STD
    counts$all<-counts$HET + counts$HOMO

     colnames(counts)<- paste0(colnames(counts),".genotype")
    counts$Inversion<-rownames(counts)
    
    test<-merge(test, counts, all=TRUE, by = "Inversion")
    test$STEP05.genotype<-"FILTERED"
    test[!is.na(test$all.genotype ) & test$all.genotype > 3 &  (test$HET.genotype > 0 & test$HOMO.genotype > 0), "STEP05.genotype"] <-"PASS"
    
    say.goods(test$STEP05.genotype)
    
    # STEP06: FINAL FILTER
    test$STEP06.final<-"FILTERED"
    test[(test$STEP04.autosomal.imputable == "PASS" & test$STEP05.genotype == "PASS" ), "STEP06.final"] <- "PASS"
    
    say.goods(test$STEP06.final)

```

Save tables

```{r}

write.csv(test, file = "report/2020-10-28_genotypeFilteringReport/inversions_filtered.csv")
write.csv(genotypes, file = "report/2020-10-28_genotypeFilteringReport/genotypes_filtered.csv")
  

``





















