<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01//EN"
"http://www.w3.org/TR/html4/strict.dtd">
<html>
	<head>
		<script type="text/javascript" src="../../js/jquery-1.4.2.min.js"></script>
		<script type="text/javascript" src="../../js/local.js"></script>
		<!-- Google Analytics tracking code-->
		<script type="text/javascript">
			var _gaq = _gaq || [];
			_gaq.push(['_setAccount', 'UA-16521993-1']);
			_gaq.push(['_trackPageview']);

			(function() {
				var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
				ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
				var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
			})();
		</script>
		<link href="../../style/css/default.css" rel="stylesheet" type="text/css">
		<title></title>
	</head>
	<body>
		<div class="header" id="header1">
			<!--<img src = "style/images/qctool.png" class = "header_img" ></img>-->
			<div class="header_text">
				qctool v2
			</div>
			<div class="header_subtext">
				A tool for quality control and analysis of gwas datasets.
			</div>
		</div>

		<section class="main_text">
			<h2>Altering variant identifying data</h2>
				<div class="task">
					<div class="task_name">
						Filling in missing chromosome information
					</div>
					<div class="task_notes">
						The <code>-assume-chromosome</code> option fills in any missing chromosome information in input data
						with the specified value.  E.g.:
					</div>
					
					<div class="task_command_line">
						$ qctool -g &lt;input file&gt; -og output.bgen -assume-chromosome &lt;chromosome&gt;
					</div>
				</div>
				<div class="task">
					<div class="task_name">
						Updating identifying data
					</div>
					<div class="task_notes">
						The <code>-map-id-data</code> option can be used to update the identifying data for each variant
						with a new set of data. E.g.:
					</div>
					<div class="task_command_line">
						$ qctool -g &lt;input file(s)&gt; -og output.bgen -map-id-data &lt;map file&gt; [+other options]
					</div>
					<div class="task_notes">
						<p>Ffor example, this might be useful when updating files to match a new genome build.</p>
						<p>The "map" file given to <code>-map-id-data</code> must be a text file
						with twelve named columns, in the following order: the current SNPID, rsid, chromosome, position, first and second alleles,
						followed by the desired updated SNPID, rsid, chromosome, position and alleles.  The first line is treated as column names
						(currently it doesn't matter what these are called.)  Variants not in this file are not affected by the mapping, and will
						be output unchanged.
					</p>
					<p>Matching of variants to the map file is controlled by the <code>-compare-variants-by</code>
						option - see the <a href="sorting.html">page on sorting data</a> for more on this option.</p>
					</div>
				</div>
				<div class="task">
					<div class="task_name">
						Aligning alleles
					</div>
					<div class="task_notes">
						<p>
						The <code>-strand</code> option can be used to update alleles and flip genotype data according
						to strand information supplied in an external file.  The general format is:
					</p>
					</div>
					<div class="task_command_line">
						$ qctool -g &lt;input file(s)&gt; -og output.bgen -strand &lt;strand file&gt; [-flip-to-match-allele &lt;column name&gt;]
					</div>
					<div class="task_notes">
						<p>The most common use of this option is to align alleles to match the forward strand of a reference sequence,
							and to flip genotypes so that the first allele is the reference allele.
						</p>
						<p>
							Strand files should have six columns which must be named as follows:
							<code>SNPID</code>,
							<code>rsid</code>,
							<code>chromosome</code>,
							<code>position</code>,
							<code>alleleA</code>,
							<code>alleleB</code>,
							<code>strand</code>, plus any additional columns.
							Strand information is read from the <code>strand</code> column.
							Alleles at variants where the strand is '+' will be processed unchanged; alleles
							at variants where the strand is '-' will be complemented (i.e. A&lt;-&gt;T, G&lt;-&gt;C);
							alleles at variants which have missing strand information - encoded as "?", or "NA", or for variants
							that are missing from the file - will be omitted from the output.
						</p>
						<p>
							The <code>-compare-variants-by</code> option controls how variants between the genotype data
							and the strand file.
							See <a href="sorting.html">the page on sorting</a> for more information.
						</p>
						<p>
							If the <code>-flip-to-match-allele</code> option is given, the strand file must contain a column
							with the specified name.  Each value in this column should be one of the two alleles of the variant.
							Alleles and genotypes are then also recoded so that the allele in the specified
							column is the first allele and the other allele the second allele.
							Note that the strand alignment is applied first - e.g. if the variant alleles are 'A' and 'G' and the
							strand is -, <code>-flip-to-match-allele</code> the column should contain 'T' or 'C'.
						</p>
					</div>
				</div>
			</div>
		</section>

		<nav class="button_bar"><div>			<div class="nav_button" name="overview">
				<a href="../../index.html">overview</a>
			</div>
			<div class="nav_button" name="documentation">
				documentation
				<ul class="menu">
					<li><a href="../../documentation/diagram_of_options.html">Diagram of options</a></li>
					<li><a href="../../documentation/alphabetical_options.html">List of options</a></li>
				</ul>
			</div>
			<div class="nav_button" name="tutorial">
				examples
				<ul class="menu">
					<li><a href="../../documentation/examples/basic.html">Basic usage</a></li>
					<li><a href="../../documentation/examples/converting.html">Converting file formats</a></li>
					<li><a href="../../documentation/examples/converting_genotypes.html">Thresholding genotypes</a></li>
					<li><a href="../../documentation/examples/sorting.html">Sorting and reordering data</a></li>
					<li><a href="../../documentation/examples/combining.html">Combining datasets</a></li>
					<li><a href="../../documentation/examples/merging.html">Merging in variants</a></li>
					<li><a href="../../documentation/examples/altering_id_data.html">Altering variant identifying data</a></li>
					<li><a href="../../documentation/examples/pipelining.html">Using qctool in a pipeline</a></li>
					<li><a href="../../documentation/examples/filtering_samples.html">Filtering samples</a></li>
					<li><a href="../../documentation/examples/filtering_variants.html">Filtering variants</a></li>
					<li><a href="../../documentation/examples/summary_statistics.html">Computing summary statistics</a></li>
					<li><a href="../../documentation/examples/annotating.html">Computing annotations</a></li>
					<li><a href="../../documentation/examples/adding_to_sample_file.html">Adding data to a sample file</a></li>
					<li><a href="../../documentation/examples/computing_ld.html">Computing LD</a></li>
					<li><a href="../../documentation/examples/computing_pcs.html">Computing principal components</a></li>
					<li><a href="../../documentation/examples/inferring_ploidy.html">Working with sex chromosomes</a></li>
				</ul>
			</div>
			<div class="nav_button" name="file_formats">
				file formats
				<ul class="menu">
					<a href="../../documentation/genotype_file_formats.html">Genotype file formats</a>
					<a href="../../documentation/sample_file_formats.html">Sample file formats</a>
					<a href="../../documentation/summary_file_formats.html">Summary statistic file formats</a>
			</div>
			<div class="nav_button" name="download">
				<a href="../../documentation/download.html">download</a>
			</div>
</div>
</nav>
		
	</body>
</html>
