<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01//EN"
"http://www.w3.org/TR/html4/strict.dtd">
<html>
	<head>
		<script type="text/javascript" src="../../js/jquery-1.4.2.min.js"></script>
		<script type="text/javascript" src="../../js/local.js"></script>
		<!-- Google Analytics tracking code-->
		<script type="text/javascript">
			var _gaq = _gaq || [];
			_gaq.push(['_setAccount', 'UA-16521993-1']);
			_gaq.push(['_trackPageview']);

			(function() {
				var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
				ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
				var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
			})();
		</script>
		<link href="../../style/css/default.css" rel="stylesheet" type="text/css">
		<title></title>
	</head>
	<body>
		<div class="header" id="header1">
			<!--<img src = "style/images/qctool.png" class = "header_img" ></img>-->
			<div class="header_text">
				qctool v2
			</div>
			<div class="header_subtext">
				A tool for quality control and analysis of gwas datasets.
			</div>
		</div>

		<section class="main_text">
					<h2>Converting between file formats</h2>
				<div class="task">
					<div class="task_name">
						Basic conversions
					</div>
					<div class="task_notes">
						The basic format of a conversion command is:
					</div>
					<div class="task_command_line">
						$ qctool -g &lt;input file&gt; [-filetype &lt;input type&gt;] -og &lt;output file&gt; [-ofiletype &lt;output type&gt;] [+modifier options]
					</div>
					<div class="task_notes">
						QCTOOL normally deduces file types from the file extension.
						However, for filetypes that are not automatically recognised, or if you want to specify,
						the <code>-filetype</code> and <code>-ofiletype</code> options can be used to specify the types.
						The <a href="../../documentation/genotype_file_formats.html">genotype file formats page</a> lists file type specifiers and
						any applicable modifier options.
					</div>
					<div class="task_notes">
						E.g. to convert between between bgen and vcf formats:
					</div>
					<div class="task_command_line">
						$ qctool -g example.bgen -og example.vcf
					</div>
				</div>
				<div class="task">
					<div class="task_name">
						Handling files split by chromosome
					</div>
					<div class="task_notes">
						If the input filename contains a # character, e.g. <code>example_#.gen</code> this is treated as a
						chromosomal wildcard and will match all (human) chromosomes. This will also be used to
						infer the chromosome for each variant if chromosome information is not present in the files themselves.
						For example, the command:
					</div>
					<div class="task_command_line">
						$ qctool -g example_#.gen -og example.bgen
					</div>
					<div class="task_notes">
						will process all 22 example files, and if the input GEN files have no chromosome information included,
						the output data will have chromosome identifiers taken from the filenames.
					</div>
				</div>
				<div class="task">
					<div class="task_name">
						Including sample information in conversions
					</div>
					<div class="task_notes">
						 Although specifying a sample file in conversions is optional, it's sometimes preferable to include one so that output files have
						the correct sample identifiers (e.g. when converting GEN format to BGEN or VCF), e.g.:
					</div>
					<div class="task_command_line">
						$ qctool -g example_#.gen -og example.bgen -s example.sample
					</div>
					<div class="task_notes">
						which will produce a BGEN file with a sample identifier block.
					</div>
				</div>
				<div class="task">
					<div class="task_name">
						Converting from VCF format
					</div>
					<div class="task_notes">
						By default QCTOOL reads genotype calls from the <span class="monospace">GT</span> field in the VCF file.
						The <code>-vcf-genotype-field</code> can be used to alter this, e.g.:
					</div>
					<div class="task_command_line">
						$ qctool -g example.vcf -vcf-genotype-field GP -og converted.bgen
					</div>
					<div class="task_notes">
						QCTOOL also assumes that VCF metadata is correct (according to the VCF v4.2 spec)- and it will
						fail with an error if that's not the case.
						To work around this, the <code>-metadata</code> option can be used to tell QCTOOL to load metadata from an
						external file:
					</div>
					<div class="task_command_line">
						$ qctool -g example.vcf -og converted.bgen -metadata metadata.txt
					</div>
					<div class="task_notes">
						The specified metadata file should contain complete VCF metadata, starting
						with the <code>##fileformat=VCFv4.2</code> line, up to but not
						including the <code>#CHROM...</code> line.
					</div>
				</div>
			</div>
		</section>

		<nav class="button_bar"><div>
			<div class="nav_button" name="overview">
				<a href="../../index.html">overview</a>
			</div>
			<div class="nav_button" name="documentation">
				documentation
				<ul class="menu">
					<li><a href="../../documentation/diagram_of_options.html">Diagram of options</a></li>
					<li><a href="../../documentation/alphabetical_options.html">List of options</a></li>
				</ul>
			</div>
			<div class="nav_button" name="tutorial">
				examples
				<ul class="menu">
					<li><a href="../../documentation/examples/basic.html">Basic usage</a></li>
					<li><a href="../../documentation/examples/converting.html">Converting file formats</a></li>
					<li><a href="../../documentation/examples/converting_genotypes.html">Thresholding genotypes</a></li>
					<li><a href="../../documentation/examples/sorting.html">Sorting and reordering data</a></li>
					<li><a href="../../documentation/examples/combining.html">Combining datasets</a></li>
<li><a href="../../documentation/examples/merging.html">Merging in variants</a></li>
					<li><a href="../../documentation/examples/altering_id_data.html">Altering variant identifying data</a></li>
					<li><a href="../../documentation/examples/pipelining.html">Using qctool in a pipeline</a></li>
					<li><a href="../../documentation/examples/filtering_samples.html">Filtering samples</a></li>
<li><a href="../../documentation/examples/filtering_variants.html">Filtering variants</a></li>
					<li><a href="../../documentation/examples/summary_statistics.html">Computing summary statistics</a></li>
					<li><a href="../../documentation/examples/annotating.html">Computing annotations</a></li>
					<li><a href="../../documentation/examples/examples/adding_to_sample_file.html">Adding data to a sample file</a></li>
<li><a href="../../documentation/examples/computing_pcs.html">Computing principal components</a></li>
					<li><a href="../../documentation/examples/inferring_ploidy.html">Working with sex chromosomes</a></li>
				</ul>
			</div>
			<div class="nav_button" name="file_formats">
				file formats
				<ul class="menu">
					<a href="../../documentation/genotype_file_formats.html">Genotype file formats</a>
					<a href="../../documentation/sample_file_formats.html">Sample file formats</a>
					<a href="../../documentation/summary_file_formats.html">Summary statistic file formats</a>
			</div>
			<div class="nav_button" name="download">
				<a href="../../documentation/download.html">download</a>
			</div>
		</div>
		
	</body>
</html>
