<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01//EN"
"http://www.w3.org/TR/html4/strict.dtd">
<html>
	<head>
		<script type="text/javascript" src="../../js/jquery-1.4.2.min.js"></script>
		<script type="text/javascript" src="../../js/local.js"></script>
		<!-- Google Analytics tracking code-->
		<script type="text/javascript">
			var _gaq = _gaq || [];
			_gaq.push(['_setAccount', 'UA-16521993-1']);
			_gaq.push(['_trackPageview']);

			(function() {
				var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
				ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
				var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
			})();
		</script>
		<link href="../../style/css/default.css" rel="stylesheet" type="text/css">
		<title></title>
	</head>
	<body>
		<div class="header" id="header1">
			<!--<img src = "style/images/qctool.png" class = "header_img" ></img>-->
			<div class="header_text">
				qctool v2
			</div>
			<div class="header_subtext">
				A tool for quality control and analysis of gwas datasets.
			</div>
		</div>

		<section class="main_text">
			<h2>Filtering variants</h2>
			<div class="task">
				<div class="task_name">
					Filtering variants based on an external file
				</div>
				<div class="task_notes">
					QCTOOL has a set of options to filter variants, each namely:
					<code>-[in|ex]cl-rsids</code>,
					<code>-[in|ex]cl-snpids</code>,
					<code>-[in|ex]cl-positions</code>,
					<code>-[in|ex]cl-variants</code>,
					<code>-[in|ex]cl-variants-matching</code>.
					Here are examples of these options:
				</div>
				<div class="task_command_line">
					$ qctool -g example.bgen -og subsetted.bgen -excl-rsids &lt;filename&gt;
				</div>
				<div class="task_notes">
					Here the specified file should contain a whitespace-separated list of rsids that will be excluded from processing.
				</div>
				<div class="task_command_line">
					$ qctool -g example.bgen -og subsetted.bgen -excl-snpids &lt;filename&gt;
				</div>
				<div class="task_notes">
					Ditto, but for alternate IDs.
				</div>
				<div class="task_command_line">
					$ qctool -g example.bgen -og subsetted.bgen -excl-positions &lt;filename&gt;
				</div>
				<div class="task_notes">
					The specified file should contain a list of genomic positions in the format <code>[chromosome:]position</code>.
					The chromosome should be omitted if you want to specify variants that have missing chromosome information.
				</div>
				<div class="task_command_line">
					$ qctool -g example.bgen -og subsetted.bgen -excl-variants &lt;filename&gt;
				</div>
				<div class="task_notes">
					The specified file should contain a list of variants.  Currently this must be a text file with six named
				 	 columns; the first four must be <code>SNPID</code>, <code>rsid</code>, <code>chromosome</code>, <code>position</code>,
					 followed by columns containing the first and second alleles.  The <code>-compare-variants-by</code> option control
					 how variants are matched to this file - see the page on <a href="sorting.html">sorting data</a> for more information
					 on this option.
				</div>
			</div>

			<div class="task">
				<div class="task_name">
					Range filtering
				</div>
				<div class="task_notes">
					The <code>-[in|ex]cl-range</code> option filters variants by range.  E.g.:
				</div>
				<div class="task_command_line">
					$ qctool -g example.bgen -og subsetted.bgen -[in|ex]cl-range [chromosome]:[start]-[end]
				</div>
				<div class="task_notes">
					This includes / excludes any variant in the given range.  Ranges are treated as closed ranges,
					i.e. the range includes both start and end positions.
					Optionally you can omit the chromosome; this will additionally capture variants that
					have missing chromosome information. Also, either the start or end positions can be omitted, in which case the range is treated as containing
					all positions up to the end coordinate, or all variants from the start coordinate onwards respectively.
					Examples of valid ranges are:
					<code>1:100-200</code>, <code>1:-200</code>, or <code>X:1000000-</code>.
				</div>
			</div>

			<div class="task">
				<div class="task_name">
					Wildcard variant filtering
				</div>
				<div class="task_notes">
					You can filter variants based on a wildcard match of ID fields.  E.g.:
				</div>
				<div class="task_command_line">
					$ qctool -g example.bgen -og subsetted.bgen -incl-variants-matching rsid~rs1%
				</div>
				<div class="task_notes">
					This command will retain all variants that have rsid starting with 'rs1'.  The general format of this command is:
				</div>
				<div class="task_command_line">
					$ qctool -g example.bgen -og subsetted.bgen -incl-variants-matching [field~][value]
				</div>
				<div class="task_notes">
					Here <code>field</code> can be 'snpid' (matching all alternate IDs) or 'rsid' (matching the first, or rs id),
					or it can be omitted to match any id.  The value can optionally contain a single '%' character, which will
					expand to match any string value.  A complete match is required, hence the value 'a%b'
					will match the ID 'ab', 'a1b', etc., but not 'zab' or 'ab2'.
				</div>
			</div>
			<div class="task">
				<div class="task_name">
					Combining multiple filters in the same command
				</div>
				<div class="task_notes">
					<p>
					The logic for processing multiple inclusion/exclusion options is as follows.
					First, if any inclusion option is specified multiple times, the results are logically ORd together.
					(Thus, for example specifying <code>-incl-range</code> twice results in including variants in either range).
					Second, the resulting conditions are ANDed together.  This means that a variant will then be
					included if it is included by each of the inclusion options and is not
					excluded by any exclusion option.
				</p>
				<p>
					For example, the following command includes any variant that is in either range and that is not in the given file:
				</p>
			</div>
				<div class="task_command_line">
					$ qctool -g example.bgen -og subsetted.bgen -incl-range 1:-1000 -incl-range 2:-1000 -excl-rsids rsids.txt
				</div>
				<div class="task_notes">
					<p>
						while the following command includes only
						variants that are in the given range and have rsid staring with "rs1":
					</p>
				</div>
				<div class="task_command_line">
					$ qctool -g example.bgen -og subsetted.bgen -incl-range 1:-1000 -incl-variants-matching rsid~rs1%
				</div>
			</div>
		</section>

		<nav class="button_bar"><div>			<div class="nav_button" name="overview">
				<a href="../../index.html">overview</a>
			</div>
			<div class="nav_button" name="documentation">
				documentation
				<ul class="menu">
					<li><a href="../../documentation/diagram_of_options.html">Diagram of options</a></li>
					<li><a href="../../documentation/alphabetical_options.html">List of options</a></li>
				</ul>
			</div>
			<div class="nav_button" name="tutorial">
				examples
				<ul class="menu">
					<li><a href="../../documentation/examples/basic.html">Basic usage</a></li>
					<li><a href="../../documentation/examples/converting.html">Converting file formats</a></li>
					<li><a href="../../documentation/examples/converting_genotypes.html">Thresholding genotypes</a></li>
					<li><a href="../../documentation/examples/sorting.html">Sorting and reordering data</a></li>
					<li><a href="../../documentation/examples/combining.html">Combining datasets</a></li>
					<li><a href="../../documentation/examples/merging.html">Merging in variants</a></li>
					<li><a href="../../documentation/examples/altering_id_data.html">Altering variant identifying data</a></li>
					<li><a href="../../documentation/examples/pipelining.html">Using qctool in a pipeline</a></li>
					<li><a href="../../documentation/examples/filtering_samples.html">Filtering samples</a></li>
					<li><a href="../../documentation/examples/filtering_variants.html">Filtering variants</a></li>
					<li><a href="../../documentation/examples/summary_statistics.html">Computing summary statistics</a></li>
					<li><a href="../../documentation/examples/annotating.html">Computing annotations</a></li>
					<li><a href="../../documentation/examples/adding_to_sample_file.html">Adding data to a sample file</a></li>
					<li><a href="../../documentation/examples/computing_ld.html">Computing LD</a></li>
					<li><a href="../../documentation/examples/computing_pcs.html">Computing principal components</a></li>
					<li><a href="../../documentation/examples/inferring_ploidy.html">Working with sex chromosomes</a></li>
				</ul>
			</div>
			<div class="nav_button" name="file_formats">
				file formats
				<ul class="menu">
					<a href="../../documentation/genotype_file_formats.html">Genotype file formats</a>
					<a href="../../documentation/sample_file_formats.html">Sample file formats</a>
					<a href="../../documentation/summary_file_formats.html">Summary statistic file formats</a>
			</div>
			<div class="nav_button" name="download">
				<a href="../../documentation/download.html">download</a>
			</div>
</div>
</nav>
		
	</body>
</html>
