<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01//EN"
"http://www.w3.org/TR/html4/strict.dtd">
<html>
	<head>
		<script type="text/javascript" src="../../js/jquery-1.4.2.min.js"></script>
		<script type="text/javascript" src="../../js/local.js"></script>
		<!-- Google Analytics tracking code-->
		<script type="text/javascript">
			var _gaq = _gaq || [];
			_gaq.push(['_setAccount', 'UA-16521993-1']);
			_gaq.push(['_trackPageview']);

			(function() {
				var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
				ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
				var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
			})();
		</script>
		<link href="../../style/css/default.css" rel="stylesheet" type="text/css">
		<title></title>
	</head>
	<body>
		<div class="header" id="header1">
			<!--<img src = "style/images/qctool.png" class = "header_img" ></img>-->
			<div class="header_text">
				qctool v2
			</div>
			<div class="header_subtext">
				A tool for quality control and analysis of gwas datasets.
			</div>
		</div>

		<section class="main_text">
			<h2>Working with sex chromosomes</h2>
				<div class="task">
					<div class="task_name">
						Handling haploid and diploid samples
					</div>
					<div class="task_notes">
						Several operations in QCTOOL, including  <code>-snp-stats</code> and <code>-compute-ld-with</code>,
						handle haploid and diploid genotypes directly.  For example, consider the command:
					</div>
					
					<div class="task_command_line">
						$ qctool -g &lt;input file&gt; -snp-stats -osnp &lt;output file&gt;
					</div>
					<div class="task_notes">
						When the input file contains X chromosome variants, the output file will contain relevant
						information including haploid and diploid genotype counts, tests for equal frequency between
						haploid and diploid samples, and appropriately estimated frequency and imputation certainty metrics.
					</div>
				</div>
			</div
			<div class="task">
				<div class="task_name">
					Inferring ploidy from sex information
				</div>
				<div class="task_notes">
					In order for this to work, it is important that haploid genotypes are encoded appropriately in input files - e.g.
					by using haploid calls in vcf GT field, or haploid data in BGEN-format files.  However, some file formats (e.g. GEN format)
					do not support haploid calls, and some datasets (E.g. the 1000 Genomes Project genotype files) choose to encode all calls
					as diploid even when the underlying genotype is haploid (i.e. for males on the X chromosome).  To work around this,
					qctool can infer the ploidy from sex information supplied in a sample file.  (This functionality currently has
					hard-coded ampping of chromosome identifiers to ploidy appropriate for humans).
				</div>
				<div class="task_notes">
					This works as follows.  Suppose <code>sex</code> is the name of a column in the sample file containing sex information
					- this must be of type 'D', and valid values are <code>male</code>, <code>M</code>, or <code>1</code> for male samples,
					and <code>female</code>, <code>F</code>, or <code>2</code> for female.  Then, the <code>-infer-ploidy-from</code> option
					can be used to tell QCTOOL to interpret diploid calls as haploid (or zero-ploid) as appropriate:
				</div>
				<div class="task_command_line">
					$ qctool -g &lt;input file&gt; -infer-ploidy-from sex -snp-stats -osnp &lt;output file&gt;
				</div>
				<div class="task_notes">
					For each genotype QCTOOL applies the following rules:
					<ol>
						<li> If input genotype is not diploid, raise an error and terminate program.</li>
						<li> If the sex of the sample is missing, set the genotype to missing.</li>
						<li> Otherwise, compute the ploidy of the sample based on the chromosome identifier and the sample sex.</li>
						<li> If the inferred ploidy is diploid, output the input genotype unchanged.</li>
						<li> If the inferred ploidy is haploid, check whether the input genotype is homozygous (or, for
						genotype probabilities, has 100% of the probability mass on homozygous calls).  If so, set the
						genotype to the corresponding haploid call.  Otherwise, set the genotype to missing.</li>
					</ol>
				</div>
				<div class="task_notes">
					These rules apply to both hard-coded genotype calls (e.g. vcf GT field) and to
					genotype probabilities (e.g. vcf GP field / GEN format / BGEN format data).
				</div>
			</div>
		</section>

		<nav class="button_bar"><div>			<div class="nav_button" name="overview">
				<a href="../../index.html">overview</a>
			</div>
			<div class="nav_button" name="documentation">
				documentation
				<ul class="menu">
					<li><a href="../../documentation/diagram_of_options.html">Diagram of options</a></li>
					<li><a href="../../documentation/alphabetical_options.html">List of options</a></li>
				</ul>
			</div>
			<div class="nav_button" name="tutorial">
				examples
				<ul class="menu">
					<li><a href="../../documentation/examples/basic.html">Basic usage</a></li>
					<li><a href="../../documentation/examples/converting.html">Converting file formats</a></li>
					<li><a href="../../documentation/examples/converting_genotypes.html">Thresholding genotypes</a></li>
					<li><a href="../../documentation/examples/sorting.html">Sorting and reordering data</a></li>
					<li><a href="../../documentation/examples/combining.html">Combining datasets</a></li>
					<li><a href="../../documentation/examples/merging.html">Merging in variants</a></li>
					<li><a href="../../documentation/examples/altering_id_data.html">Altering variant identifying data</a></li>
					<li><a href="../../documentation/examples/pipelining.html">Using qctool in a pipeline</a></li>
					<li><a href="../../documentation/examples/filtering_samples.html">Filtering samples</a></li>
					<li><a href="../../documentation/examples/filtering_variants.html">Filtering variants</a></li>
					<li><a href="../../documentation/examples/summary_statistics.html">Computing summary statistics</a></li>
					<li><a href="../../documentation/examples/annotating.html">Computing annotations</a></li>
					<li><a href="../../documentation/examples/adding_to_sample_file.html">Adding data to a sample file</a></li>
					<li><a href="../../documentation/examples/computing_ld.html">Computing LD</a></li>
					<li><a href="../../documentation/examples/computing_pcs.html">Computing principal components</a></li>
					<li><a href="../../documentation/examples/inferring_ploidy.html">Working with sex chromosomes</a></li>
				</ul>
			</div>
			<div class="nav_button" name="file_formats">
				file formats
				<ul class="menu">
					<a href="../../documentation/genotype_file_formats.html">Genotype file formats</a>
					<a href="../../documentation/sample_file_formats.html">Sample file formats</a>
					<a href="../../documentation/summary_file_formats.html">Summary statistic file formats</a>
			</div>
			<div class="nav_button" name="download">
				<a href="../../documentation/download.html">download</a>
			</div>
</div>
</nav>
		
	</body>
</html>
